--- 
title: "Differential gene expression analyses"
output: 2018-01-09-DE.nb.html
--- 
---
### Introduction 
#This script is for differential gene expression analysis of transcript quantification data.
#Input files should be abundance files (.h5, .tsv, .json) generated by Kallisto (https://pachterlab.github.io/kallisto/manual) for each sample. 
#This script will import abundance files from Kallisto and use the R package 'DESeq2' (http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for differential expression based on a model using negative binomial distribution.
#This script produces a results table of genes that are significantly differentially expressed, following DESeq2 analysis. The results of which can then be used for further analysis. 
---

```{r}
########## load packages 
library(tximport)
library(readr)
library(DESeq2)
library(ggplot2)
library(ggfortify)
library(ReportingTools)
source("https://bioconductor.org/biocLite.R")
biocLite("vsn")
library("vsn")
library("RColorBrewer")

### set wd 
setwd("/Users/isabelfletcher/Documents/LIDo/R1/RNAseq/2018-01-15-Differential_Expression")

```
##Step One:
## Input results from kallisto
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$data <- "." #confusing
paths$kallisto_output <- file.path(paths$data, "2018-01-11-kallisto_quantification")

paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", value = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)


# Save path to use later
kallisto_quant_file_path <- "2018-01-11-kallisto_quantification"

# Automatically extract file names from kallisto output for colnames
names_tmp <- gsub(x = paths$kallisto_files, pattern = "./2018-01-11-kallisto_quantification/2016-Bter-", replacement = "")
names_tmp <- gsub(x = names_tmp, pattern = "-head.+", replacement = "")
names <- gsub(x = names_tmp, pattern = "-.+", replacement = "")
rm(names_tmp)

```
##Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport
```{r}
## Extract sample names
samples_tmp <- gsub(x = paths$kallisto_files, pattern = ".*Bter-", replacement = "")
samples     <- gsub(x = samples_tmp,           pattern = "-.*",     replacement = "")
samples  <- data.frame(treatment = samples)
rownames(samples) <- gsub(x = samples_tmp,     pattern = "-head.*",     replacement = "")
rm(samples_tmp)

```
##Step Three: Tximport
```{r}
## Use tximport on Kallisto files 
txi <- tximport(paths$kallisto_files, type = "kallisto", txOut = TRUE)

colnames(txi$abundance) <- names
head(txi$abundance)

# Save object for use in other scripts
save(txi, file = "txi.Rdata")

```
##Step Four:
## Run DESeq2 for differential expression analysis 
```{r}
############## DESeq
# Construct DESeq data set from txi object and sample information
deseq_txi <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ treatment) 

# Perform differential expression analysis
deseq_object <- DESeq(deseq_txi)

```

```{r}
#### Generate a results table for DE, specifying contrasts as treatment CON vs. CLO
deseq_results <- results(deseq_object, contrast = c("treatment", sort(levels(samples$treatment), 
                                              decreasing = TRUE))) 
# Save as object for further use
save(deseq_object, file = "deseq_object.Rdata")

## Log fold change shrinkage for ranking and visualisation, provide lfcshrink() with the DESeq object 
deseq_results_names <- resultsNames(deseq_object)
deseq_results_shrink <- lfcShrink(deseq_object, coef = deseq_results_names[2]) 

# Order lfcshrink results by p value
deseq_results_ordered <- deseq_results_shrink[order(deseq_results_shrink$pvalue),]

# Number of significantly differentially expressed genes 
nrow(deseq_results_ordered[which(deseq_results_ordered$padj < 0.05), ])
                                              
```

```{r}
## Create table of significant genes 
deseq_results_significant <- deseq_results_ordered[which(deseq_results_ordered$padj < 0.05), ]

# Save table as R object for use in other scripts 
save(deseq_results_significant, file = "deseq_results_significant.Rdata")

# Put significant gene transcripts into an object 
de_genes <- rownames(deseq_results_significant) 
de_genes <- gsub(x = de_genes, pattern = ".+rna_", replacement = "")
de_genes <- gsub(x = de_genes, pattern = "[:.:][0-9].+", replacement = "")

# Export object into .txt file, for GO analysis
write(de_genes, "de_output.txt")
```

```{r}
## Table of results based on log fold changes in expression
deseq_results_logchange <- deseq_results_significant[order(deseq_results_significant$log2FoldChange),]

# Trim rownnames
rownames(deseq_results_logchange) <- gsub(x = rownames(deseq_results_logchange), pattern = ".+rna_", replacement = "")
rownames(deseq_results_logchange) <- gsub(x = rownames(deseq_results_logchange), pattern = "[:.:][0-9].+", replacement = "")
de_genes_logfold <- rownames(deseq_results_logchange)

# Export object
write(de_genes_logfold, "de_output_logfold.txt")

```

##Step Five
```{r}
## Select significant gene abundances, for use in other scripts
# Select significant abundances
if ( any(rownames(txi$abundance) != rownames(deseq_results_shrink))) {
  stop("we have a problem with rownames")
}
sig_gene_abundances <- txi$abundance[which(deseq_results_shrink$padj < 0.05),] 

# Create object for use in later script
save(sig_gene_abundances, file = "sig_gene_abundances.Rdata")
```

##Step Six: Principal Components Analysis
```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object, blind = FALSE)
rld <- rlog(deseq_object, blind = FALSE)

# this gives log2(n + 1)
ntd        <- normTransform(deseq_object)

## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))

## Extract first and second principal components:
pca_data    <- plotPCA(vsd, intgroup = c("treatment"), returnData = TRUE)

## Extract percentages for first two principal components:
percentVar <- round(100 * attr(pca_data, "percentVar"))

## Plot PCA:
ggplot(pca_data, aes(PC1, PC2, color = treatment, shape = treatment)) +
       geom_point(size = 5, shape = 16) +
       xlab(paste0("PC1: ",percentVar[1],"% variance")) +
       ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
       coord_fixed() +
       scale_color_manual(values = c("blue", "grey")) +
       theme_bw() +
       theme(axis.title = element_text(family = "Arial", color = "#666666", face = "bold", size = 16)) 

        
```


#Step Six
## Making report with boxplots, for exploratory analysis 
```{r}
deseq_report <- HTMLReport(shortName = 'RNAseq_analysis_with_DESeq2',
                         title = 'RNA-seq analysis of differential expression using DESeq2',
                         reportDirectory = "./reports")

publish(deseq_object,deseq_report, pvalueCutoff = 0.05,
        factor = colData(deseq_object)$treatment,
        reportDir = "./reports")
        
finish(deseq_report)
```






