## Context & citation

This repository contains all scripts used for the differential expression analysis of _Bombus terrestris_ workers and queens in response to neonicotinoid exposure. This was developed as part of a NERC-funded project reported implemented in [Yannick Wurm's lab at Queen Mary University of London](http://wurmlab.com). Findings are reported in the following manuscript: 

Colgan, T.J., Fletcher, I.K., Arce, A.N., Gill, R.J., Ramos Rodrigues, A., Stolle, E., Chittka, L. and Wurm, Y.  
**Caste- and pesticide-specific effects of neonicotinoid pesticide exposure on gene expression in bumblebees**. *in prep*.  

## License

This work is distributed under the [GPLv3 license](LICENSE). Reuse of code derived from this repository is permitted under two conditions: 
1) Proper attribution (i.e., citation of the associated publication; see [CITATION.cff](CITATION.cff) and above).
2) Publication of reused scripts on an open-access platform, such as Github.  

## Workflow

Analysis workflow consisted of:  
1. Quality assessment of aligned data using [multiQC](https://multiqc.info/).  
2. Pseudoalignment and transcript quantification using [kallisto](https://pachterlab.github.io/kallisto/).     
3. Splice-aware alignment using [HISAT2](https://ccb.jhu.edu/software/hisat2/index.shtml).  
4. Exon-level read count generation using [HTSeq](https://htseq.readthedocs.io/).      
5. Differential expression analysis using [DESeq2](https://www.bioconductor.org/packages/release/bioc/html/DESeq2.html).    
6. Differential exon usage analysis using [DEXSeq](https://bioconductor.org/packages/release/bioc/html/DEXSeq.html).     
7. Gene Ontology enrichment analysis using [topGO](https://bioconductor.org/packages/release/bioc/html/topGO.html).   
8. Comparative analysis of read counts generated by kallisto and HTSeq.  
9. Generation of graphics for visualisation of results.  

### Repository contents:  
```
01_multiqc_quality_assessment
|- hisat2_to_multiQC.sh
02_amplitude_kallisto_quantifications/
|- run_kallisto.sh
03_spliced_mapping_hisat
|- hisat_mapping.sh
04_exon_level_read_count_quantifications
|- prepare_input_files_for_dexseq.sh
05_differential_expression_analysis
|- treatment_caste_group_analysis  
|- |- gene_level_group_analysis.Rmd
|- treatment_caste_interaction_analysis
|- |- gene_level_3_treatments.Rmd
|- |- input  
|- |- |- sample_information.txt
|- treatment_specific_analysis
|- |- gene_level_3_treatments.Rmd
|- |- input  
|- |- |- sample_information.txt
06_differential_exon_usage_analysis
|- dexseq_analysis.Rmd
|- get_lowest_exon_p_value.sh
07_gene_ontology_enrichment_analysis
|- go_enrichment_analysis.Rmd
|- input
|- |- dmel_vs_bter_biomart.input_for_converter.output.txt
08_comparative_read_count_analysis
|- compare_kallisto_hisat2_quants.Rmd
09_plot_visualisation
|- euler_diagram.Rmd
|- plot_DEGs_heatmap.Rmd
|- plot_go_barchart.Rmd
utility
|- barchart_helper_functions.R
|- description_from_LOCid.R
|- helper_functions_for_heatmaps.R
```
