--- 
title: "Bombus pesticide exposure: gene expression in head"
output: compare_kallisto_hisat2_quants.html
author: Joe Colgan http://wurmlab.com
--- 

## Introduction  
This scripts takes in the output quantification files generated by both Kallisto and HISAT2-HTSeq and compares expression across genes to investigate if there is strong correlation between the gene-level quantifications generated by both approaches.  

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2",
               "ggplot2", "ggpubr", "rhdf5",
               "plyr", "ggfortify", "biomaRt",
               "vsn", "RColorBrewer", "ReportingTools")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        library(lib, character.only = TRUE )
    }
}
## Create output directory:
dir.create("results")
```

## Read in hisat2-htseq counts:

```{r, message=FALSE}
## Read in files:
## a) Data for queens:
hisat_htseq_q <- read.table(file = "./final_queen_hisat2_htseq_counts.txt",
                          header = TRUE)

## Append column names:
colnames(hisat_htseq_q) <- paste(colnames(hisat_htseq_q), "Q", sep = "_")

## b) Data for workers:
hisat_htseq_w <- read.table(file = "./final_worker_hisat2_htseq_counts.txt",
                          header = TRUE)

## Append column names:
colnames(hisat_htseq_w) <- paste(colnames(hisat_htseq_w), "W", sep = "_")

## Remove gene id column for queen data and combine worker and queen to make a combined dataframe:
hisat_htseq_q$gene_id_Q <- NULL
hisat_htseq_counts_all <- cbind(hisat_htseq_w, hisat_htseq_q)

rownames(hisat_htseq_counts_all) <- hisat_htseq_counts_all$gene_id_W
hisat_htseq_counts_all$gene_id_W <- NULL
```

## Compare with kallisto counts:
## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.

```{r, message=FALSE} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$sample_batch_info <- "room_batch_information_per_sample-clean.txt"
paths$kallisto_output <- "./input"

# Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output,
                                                recursive = TRUE),
                                      pattern = "h5",
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
names(paths$kallisto_files) <- gsub(paths$kallisto_files_relative,
                                    pattern = "/.*", replacement = "")
for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```

## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples     <- data.frame(sample_name = names(paths$kallisto_files))
## Read in sample information for room and batch:
samples_information <- read.table(file = paths$sample_batch_info,
                                  header = FALSE,
                                  col.names = c("sample_name",
                                                "treatment",
                                                "room", "batch"),
                                  row.names = 1)
## Make sure 'CON' is the reference:
samples_information$treatment <- relevel(x = samples_information$treatment,
                                         ref = "CON")
# sanity check
if (FALSE == all(samples$treatment %in% rownames(samples_information))) {
    kill("we have a problem - sample names from files and from
         batch descriptor file  dont match up")
}
samples_information
```

## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
ensembl <- useMart(biomart = "metazoa_mart",
                  host = "jul2018-metazoa.ensembl.org",
                  dataset = "bterrestris_eg_gene")

transcript_to_gene <- getBM(attributes = c("ensembl_transcript_id",
                                           "ensembl_gene_id"),
                            mart = ensembl)

# Use tximport on kallisto files
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files,
                       type    = "kallisto",
                       tx2gene = transcript_to_gene,
                       countsFromAbundance = "no")

## Count the number of genes per input data.frame:
nrow(txi_counts$counts)
nrow(hisat_htseq_counts_all)

## Remove tags:
MetaTags <- grep("^__", rownames(hisat_htseq_counts_all))
hisat_htseq_counts_all <- hisat_htseq_counts_all[-MetaTags, ]

## Match and extract only genes found in both hisat and kallisto counts:
txi_counts_df <- as.data.frame(txi_counts$counts[match(rownames(hisat_htseq_counts_all),
                                                       rownames(txi_counts$counts)), ])

## Calculate rowsums for each gene within each dataset
txi_counts_df$mean <- rowMeans(txi_counts_df)
hisat_htseq_counts_all$mean <- rowMeans(hisat_htseq_counts_all)

## Combine rowsums:
combined.df <- as.data.frame(cbind(txi_counts_df$mean,
                                   hisat_htseq_counts_all$mean))
rownames(combined.df) <- rownames(txi_counts_df)

## Update column names:
colnames(combined.df) <- c("kallisto", "hisat_htseq")

## Check correlation:
cor.test(combined.df$kallisto, combined.df$hisat_htseq)

## Remove low counts:
combined.df <- subset(combined.df, kallisto > 10 &
                             hisat_htseq > 10)

## For the plots, turn off scientific notation:
options(scipen = 999)

## Generate plots:
## a) Normal plot:
normal_plot <- ggplot(data = combined.df, aes(x = kallisto,
                                       y = hisat_htseq)) +
                geom_point(alpha = 0.3, size = 4) +
                xlab("Kallisto raw gene-level counts") +
                ylab("HISAT2-HTSeq raw gene-level counts") +
                theme_bw()

## Increase font size for text labels:
normal_plot <- normal_plot + theme(axis.title = element_text(face = "bold",
                                               size = "25"),
                                   axis.text.x = element_text(size = "20",
                                                              angle = 45,
                                                              hjust = 1),
                                   axis.text.y = element_text(size = "20"))

## Print to console:
normal_plot

## Save:
ggsave(file = "correlation_plots_mean_raw_estimated_counts.png",
       height = 10, width = 20)


## b) Generate a plot using log_transformed values:
log_plot <- ggplot(data = combined.df, aes(x = kallisto,
                                       y = hisat_htseq)) +
        geom_point(alpha = 0.3, size = 4) +
        xlab("Kallisto log2 transformed gene-level counts") +
        ylab("HISAT2-HTSeq log2 transformed gene-level counts") +
        coord_trans(x = "log2", y = "log2") +
        theme_bw()
        
## Increase font size for text labels:
log_plot <- log_plot + theme(axis.title = element_text(face = "bold",
                                               size = "25"),
                             axis.text.x = element_text(size = "18",
                                                        angle = 45,
                                                        hjust = 1),
                             axis.text.y = element_text(size = "18"))

## Print to console:
log_plot

## Save:
ggsave(file = "correlation_plots_mean_log_estimated_counts.png",
       height = 10, width = 20)

## Plot together:
ggarrange(normal_plot,
          log_plot,
          ncol = 2, nrow = 1,
          align = "hv",
          labels = c("A", "B"))

## Save:
ggsave(file = "two_correlation_plots_estimated_counts.png",
       height = 10, width = 20)

## Option two for plotting:
ggarrange(normal_plot,
          log_plot,
          ncol = 1, nrow = 2,
          align = "hv",
          labels = c("A", "B"))

## Save:
ggsave(file = "two_correlation_plots_estimated_counts_inverse.png",
       height = 12, width = 20)

## Generate a plot using raw and log_transformed values for a random sample:
random_sample_kallisto <- sample(colnames(txi_counts_df), 1)
random_sample_hisat <- gsub("-", ".", random_sample_kallisto)

## Extract and combine into new dataframe for plotting:
random_sample_df <- as.data.frame(cbind(txi_counts_df[[random_sample_kallisto]],
                                 hisat_htseq_counts_all[[random_sample_hisat]]))

## Update column names:
colnames(random_sample_df) <- c("kallisto", "hisat_htseq")

## Remove low counts:
random_sample_df <- subset(random_sample_df,
                           kallisto > 10 &
                           hisat_htseq > 10)

## c) Generate a plot using raw estmated counts for a random sample:
random_normal_plot <- ggplot(data = random_sample_df,
                             aes(x = kallisto,
                                 y = hisat_htseq)) +
                        geom_point(alpha = 0.3, size = 4) +
                        xlab("Kallisto raw gene-level counts") +
                        ylab("HISAT2-HTSeq raw gene-level counts") +
                        #geom_smooth(method = "lm", formula = y~x) +
                        theme_bw()

## Increase font size for text labels:
random_normal_plot <- random_normal_plot +
        theme(axis.title = element_text(face = "bold",
                                        size = "25"),
              axis.text.x = element_text(size = "17",
                                         angle = 45,
                                         hjust = 1),
              axis.text.y = element_text(size = "17"))

## Print to console:
random_normal_plot

## Save:
ggsave(file = "correlation_plots_random_raw_estimated_counts.png",
       height = 10, width = 20)

## d) Generate a log_transformed plot:
random_log_plot <- ggplot(data = random_sample_df,
                             aes(x = kallisto,
                                 y = hisat_htseq)) +
                        geom_point(alpha = 0.3, size = 4) +
                        xlab("Kallisto log2 transformed gene-level counts") +
                        ylab("HISAT2-HTSeq log2 transformed gene-level counts") +
                        coord_trans(x = "log2", y = "log2") +
                        theme_bw()

## Increase font size for text labels:
random_log_plot <- random_log_plot +
                   theme(axis.title = element_text(face = "bold",
                                               size = "25"),
                         axis.text.x = element_text(size = "17",
                                                    angle = 45,
                                                    hjust = 1),
                         axis.text.y = element_text(size = "17"))
## Print to console:
random_log_plot

## Save:
ggsave(file = "correlation_plots_random_log_estimated_counts.png",
       height = 10, width = 20)

## Plot all four plot:
ggarrange(normal_plot,
          log_plot,
          random_normal_plot,
          random_log_plot,
          ncol = 2, nrow = 2,
          align = "hv",
          labels = c("A", "B", "C", "D"))

## Save to file:
ggsave(file = "four_correlation_plots_estimated_counts.png",
       height = 15, width = 25)

## Problem with import to Google Docs, figure may be too big:
ggarrange(random_normal_plot,
          random_log_plot,
          ncol = 2, nrow = 1,
          align = "hv",
          labels = c("C", "D"))

## Save:
ggsave(file = "two_correlation_plots_random_estimated_counts.png",
       height = 10, width = 20)

```

Explore correlation between kallisto counts and htseq counts for indvidual samples:

```{r, message = FALSE}
## Combine kallisto and htseq counts:
all_samples_df <- cbind(txi_counts_df, hisat_htseq_counts_all)

## An empty list:
correlation_list <- list()

## Calculate Pearson's r2 for kallisto and htseq counts for each sample:
for (name in colnames(txi_counts_df)){
        ## Print to console the column name:
        print(name)
        ## Convert kallisto column name into htseq column name:
        new_name <- gsub("-", ".", name)
        ## Print new name:
        print(new_name)
        ## Add correlation values to an empty dataframe:
        correlation_list[[name]] <- cor(all_samples_df[[name]],
                                        all_samples_df[[new_name]])
}

## Summary of correlations:
summary(as.numeric(unlist(correlation_list)))
```
