--- 
title: "Bombus pesticide exposure: gene expression in head"
output: compare_kallisto_hisat2_quants.html
--- 

# Introduction  
This scripts takes in the output quantification files generated by both Kallisto and HISAT2-HTSeq and compares expression across genes to investigate if there is strong correlation between the gene-level quantifications generated by both approaches.  

```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2", "ggplot2", "gplots", "rhdf5",
               "plyr", "ggfortify", "biomaRt", "vsn", "RColorBrewer", "ReportingTools")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
#        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
## Create output directory:
dir.create("results")
```


## Read in hisat2-htseq counts:

```{r, message=FALSE}
## Read in file:
hisat_htseq <- read.table(file="./results/hisat2_htseq_quants_output/final_sample_list.txt",
                          header = TRUE)

rownames(hisat_htseq) <- hisat_htseq$gene_id
hisat_htseq$gene_id <- NULL
```

## Compare with kallisto quants:

```{r, message=FALSE}
## Load counts from kallisto:
## Set paths to folders containing output files from kallisto
paths <- list()
paths$kallisto_output <- "kallisto_quantifications"

# Set relative paths:
paths$kallisto_files_relative <- grep(x       = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", 
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
filenames_tmp <- paths$kallisto_files
filenames_tmp <- gsub(x = filenames_tmp, pattern     = file.path(paths$kallisto_output, "2016-Bter-"), 
                  replacement = "")
filenames_tmp <- gsub(x = filenames_tmp, pattern = "-head.+", replacement = "")
filenames     <- gsub(x = filenames_tmp, pattern = ".*Bter-", replacement = "")
names(paths$kallisto_files) <- filenames 

for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
rm(filenames_tmp, filenames)
```

## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.

```{r, message=FALSE} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$sample_batch_info <- "input/room_batch_information_per_sample-clean.txt"
paths$kallisto_output <- "kallisto_quantifications/"
# Set relative paths:
paths$kallisto_files_relative <- grep(x       = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", 
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)
# Automatically extract file names from kallisto output for colnames
names(paths$kallisto_files) <- gsub(paths$kallisto_files_relative, pattern="/.*", replacement="") 
for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```

## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples     <- data.frame(sample_name = names(paths$kallisto_files))
## Read in sample information for room and batch:
samples_information <- read.table(file = paths$sample_batch_info,
                                  header = FALSE, 
                                  col.names = c("sample_name", "treatment", "room", "batch"),
                                  row.names = 1)
## Make sure 'CON' is the reference:
samples_information$treatment <- relevel(x = samples_information$treatment, ref = "CON") 
# sanity check
if (FALSE == all(samples$treatment %in% rownames(samples_information))) {
    kill("we have a problem - sample names from files and from batch descriptor file  dont match up")
}
samples_information
```

## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
ensembl <- useMart("metazoa_mart", host = "metazoa.ensembl.org")
ensembl <- useDataset("bterrestris_eg_gene", mart = ensembl)
transcript_to_gene <- getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id"), 
                            mart = ensembl)
# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files, 
                       type    = "kallisto", 
                       tx2gene = transcript_to_gene,
                       countsFromAbundance = "no")
## 
nrow(txi_counts$counts)
nrow(hisat_htseq)

## Match and extract only genes found in both hisat and kallisto counts:
txi_counts_df <- as.data.frame(txi_counts$counts[match(rownames(hisat_htseq), rownames(txi_counts$counts)), ])

## Calculate rowsums for each gene within each dataset
txi_counts_df$mean <- rowMeans(txi_counts_df)
hisat_htseq$mean <- rowMeans(hisat_htseq)

## Combine rowsums:
combined.df <- as.data.frame(cbind(txi_counts_df$mean, hisat_htseq$mean))

colnames(combined.df) <- c("kallisto", "hisat_htseq")

## Check correlation:
cor.test(combined.df$kallisto, combined.df$hisat_htseq)
```