---
title: "Bombus pesticide exposure: gene expression in head"
output: GO_enrichment.html
---

# Introduction 
This script is for gene ontology (GO) enrichment analysis of differentiallly expressed genes to explore the biological processes and molecular functions associated with DE genes, using the R package 'TopGO' (https://www.bioconductor.org/packages/3.7/bioc/vignettes/topGO/inst/doc/topGO.pdf).
Twp input files are required for the running of the test:  
1) A genelist file:  
The genelist file is a tab-delimited file containing two columns:  
Column 1: Locus (contains gene or transcript name of interest).  
Column 2: Rank value of interest (e.g. p-values or log fold changes).  
Rank values were generated using the R package, DESeq2, and implemented using code in ```gene_level_3_treatments.Rmd```.  

2) A GO database file:  
The GO database file is a tab-delimited file containing two columns:  
Column 1: Locus (contains gene or transcript name of interest).  
Column 2: Comma separated GO terms (e.g. GO:0000001, GO:0000002, etc.).  

This script will prepare data for GO analysis and create a 'TopGO object' from which enrichment tests can be performed to explore GO terms significantly enriched within the dataset. 
This script outputs a results table of significantly enriched GO terms.

```{r}
# Load libraries; install from scratch if needed
libraries <- c("topGO", "lintr")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
```

## Step One: Input files, define output and objects for running topGO: 
Load in genelist and database files.  

```{r}
## Define input (i.e. genelist) file:
input <- "de_genes_for_topGO_CLO.txt"

treatment <- "CLO"
output_directory <- "output"
if file.exists(output_directory) { 
  stop("The output directory:", output_directory, ", already exists", 
       "Let's avoid overwriting")
} else {
  dir.create(output_directory) 
}

## Read in input file:
genes_interest <- read.table(input, header = TRUE)

## Sort by adjusted p-value:
## N.B. KS test requires ranked values - this step is to ensure values are ranked:  
genes_of_interest_sorted <- genes_of_interest[order(genes_of_interest$adjusted_pvalue), ]

## Remove na values:
## NA values are present for certain genes without expression, such as tRNAs:
genes_of_interest_sorted_filtered <- subset(x      = genes_of_interest_sorted, 
                                            subset = !is.na(adjusted_pvalue))

## Read in GO annotations:  
gene_id2go <- readMappings(file = "input/blast2go_mapping_20171223_1007.input_database_for_topGO.txt")

## Define gene 'universe':  
gene_universe <- names(gene_id2go)

## Create ranked and filtered genelist (i.e genes of interest):  
genelist        <- genes_of_interest_sorted_filtered$adjusted_pvalue
names(genelist) <- genes_of_interest_sorted_filtered$locus

# Define a cut-off for running fisher's exact test:  
cutoff_for_top_fivepercent <- quantile(x = genelist, probs = 0.05)
```

## Steps Two and Three: Create topGO Object & run tests for GO term enrichment

We create a topGO object for each GO term

We perform two statistical tests:
1. A ks test using the topGO 'weight01' algorithm
2. A Fisher's exact test using the topGO 'weight01' algoritm

We combine the output of each test. 
We filter out enriched terms.

We do this for each of the three GO categories (ie. Biological process, Molcular Function, Cellular Component):


```{r}
for (go_category in c("BP", "MF", "CC") {
  ## Build the GOdata object in topGO
  my_go_data <- new("topGOdata",
                    description = paste("GOtest", go_category, sep="_"),
                    ontology    = go_category,
                    geneSel     = function(x) { # fails to run without this
                                     return(x < cutoff_for_top_fivepercent)
                                   },
                    allGenes    = genelist,
                    gene2GO     = gene_id2go,
                    annot       = annFUN.gene2GO,
                    nodeSize    = 100) # Modify to reduce/increase stringency.
  
  ## Calculate ks test using 'weight01' algorithm:
  result_weight_ks <- runTest(object    = my_go_data,
                              algorithm = "weight01",
                              statistic = "ks")
  
  ## Calculate fisher exact test using 'weight01' algorithm:
  result_weight_fisher <- runTest(object    = my_go_data,
                                  algorithm = "weight01",
                                  statistic = "fisher")
  
  ## Combine results from statistical tests:
  result_weight_output <- GenTable(object       = my_go_data,
                                   weight_ks     = result_weight_ks,
                                   weight_fisher = result_weight_fisher,
                                   orderBy       = "weight_ks",
                                   topNodes      = length(score(result_weight_ks)))
  
  ## Subset calls with significance higher than expected:
  result_weight_output_sig <- subset(x      = result_weight_output, 
                                     subset = (Significant > Expected) & (weight_ks < 0.05))
  
  ## Print to console one of the GO terms of interest to check the distribution of that GO term across ranked genes:
  print(showGroupDensity(object  = my_go_data,
                         whichGO = head(result_weight_output_sig$GO.ID, n = 1),
                         ranks   = TRUE,
                         rm.one  = FALSE))
  
  ## Write to output:
  write.table(x         = result_weight_output_sig,
              file      = file.path(output_directory,
                                    paste(treatment, go_category, "sig.tsv", sep="_"))
              row.names = TRUE,
              sep       = "\t")
}
```
