---
title: "Bombus pesticide exposure: gene expression in head"
output: GO_enrichment.html
---

# Introduction 
This script is for gene ontology (GO) enrichment analysis of differentiallly expressed genes to explore the biological processes and molecular functions associated with DE genes, using the R package 'TopGO' (https://www.bioconductor.org/packages/3.7/bioc/vignettes/topGO/inst/doc/topGO.pdf).
Twp input files are required for the running of the test:  
1) A genelist file:  
The genelist file is a tab-delimited file containing two columns:  
Column 1: Locus (contains gene or transcript name of interest).  
Column 2: Rank value of interest (e.g. p-values or log fold changes).  
Rank values were generated using the R package, DESeq2, and implemented using code in ```gene_level_3_treatments.Rmd```.  

2) A GO database file.
The GO database file is a tab-delimited file containing two columns:  
Column 1: Locus (contains gene or transcript name of interest).  
Column 2: Comma separated GO terms (e.g. GO:0000001, GO:0000002, etc.).  

This script will prepare data for GO analysis and create a 'TopGO object' from which enrichment tests can be performed to explore GO terms significantly enriched within the dataset. 
This script outputs a results table of significantly enriched GO terms.

```{r}
# Load libraries; install from scratch if needed
libraries <- c("topGO")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
```

## Step One: Input files and define objects for running topGO: 
Load in genelist and database files.  

```{r}
## Define input (i.e. genelist) file:
input <- "de_genes_for_topGO_LRT.txt" 

## Read in input file:
genes_of_interest <- read.table(input, header = TRUE)

## Sort by adjusted p-value:
## N.B. KS test requires ranked values - this step is to ensure values are ranked:  
genes_of_interest_sorted <- genes_of_interest[order(genes_of_interest$pvalue), ]

## Remove na values:
## I need to check this but there appeared to be 'NA' values in the last DE output I used:  
genes_of_interest_sorted_filtered <- subset(genes_of_interest_sorted, subset = !is.na(pvalue))

## Read in GO annotations:  
geneID2GO <- readMappings(file = "input/blast2go_mapping_20171223_1007.input_database_for_topGO.txt")

## Define gene 'universe':  
gene_universe <- names(geneID2GO) 

## Create ranked and filtered genelist (i.e genes of interest):  
geneList        <- genes_of_interest_sorted_filtered$pvalue
names(geneList) <- genes_of_interest_sorted_filtered$locus

# Define a cut-off for running fisher's exact test:  
cutoff_for_top_fivepercent <- quantile(x = geneList, probs = 0.05)
```

## Step Two: Create topGO data object, using BP for 'Biological Processes'
We create a topGO object using GO terms associated with 'Biological Processes'

```{r}
## Variables to define - 
## Biological process - 'BP' 
## Molecular function - 'MF' 
## Cellular component - 'CC'
GO_term_type <- 'BP'

## Build the GOdata object in topGO
myGOdata <- new("topGOdata", 
                description = "My project", 
                ontology = GO_term_type,
                ## To use all genes for statistical tests with no predefined genes of interest:
                #geneSel = function(x) {return(x=TRUE)}, # Run just for ks test
                ## To create a subset of genes of interest: 
                geneSel = function(x) {return(x < cutoff_for_top_fivepercent)},
                allGenes = geneList,  
                gene2GO = geneID2GO, 
                annot = annFUN.gene2GO, 
                nodeSize=10) # NodeSize can be reduced or increased to reduce/increase stringency. 
```

## Step Three: Run statistical tests for GO term enrichment
We perform two statistical tests:
1. A ks test using the topGO 'weight01' algorithm
2. A Fisher's exact test using the topGO 'weight01' algoritm

We combine the output of each test. 
We filter out enriched terms.

a) For all three GO categories (ie. BP, MF, CC):
```{r}
## Calculate ks test using 'weight01' algorithm:
resultWeightKs <- runTest(myGOdata, 
                          algorithm = "weight01", 
                          statistic = "ks")

## Calculate fisher exact test using 'weight01' algorithm:
resultWeightFisher <- runTest(myGOdata, 
                              algorithm = "weight01", 
                              statistic = "fisher")

## Combine results from statistical tests:
allRes <- GenTable(myGOdata, 
                   weight = resultWeightKs, 
                   weightfisher = resultWeightFisher,
                   orderBy = "weight", 
                   topNodes = 2038) ## This number should be altered based on the ontology type examined.
                                    ## If number is too high, the console will complain and print out the actual
                                    ## number of nodes allowed to examine.
                                    ## If number is too low, it won't complain but you want to make sure you are using
                                    ## all of the nodes. In this scenario, increase the number of topNodes value until
                                    ## you get a complaint and then use the number printed to console. 

## Filter out calls with significance higher than expected:
allRes_sig <- allRes[which(allRes$Significant > allRes$Expected &
                           allRes$weight < 0.05 &
                           allRes$weightfisher < 0.05), ]

## Print to console one of the GO terms of interest to check the distribution of that GO term across ranked genes:
print(showGroupDensity(myGOdata_3_treatments, 
                       "GO:0044707", 
                       ranks = TRUE, 
                       rm.one = FALSE))

## Write to output:
write.table(allRes_sig, file = "allRes_sig_GO_results.txt", row.names = TRUE, sep = "\t")
```
