--- 
title: "Graphical representation of DE genes"
--- 

## Introduction:
#=================
##This script is for exploratory analysis of transcript quantification data and results of differential gene expression analysis.
##Input files should be R objects saved and generated from previous scripts of DE analysis, including the object resulting from using tximport() on kallisto files, the significant transcript abundances selected from this object and a results table of significant genes, generated using DESeq2. 
##This script will organise input data for generating heatmaps of significantly differentially expressed genes, based on statistical significance and log fold changes in expression.
##This script produces heatmaps that can be exported. 

```{r}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2", "gplots", "heatmap3", "ggfortify", "ggDESeq",
               "reshape2", "scales")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}

# Load R objects from previous script output
load("input/graphical_representation/txi_counts.Rdata")
load("input/graphical_representation/deseq_object.Rdata")
load("input/graphical_representation/de_genes_CLO.RData")
load("input/graphical_representation/de_genes_IMI.RData")
load("input/graphical_representation/de_genes_lfc.RData")

dir.create("results/graphical_representation")

```

## Step One: Heatmaps for visualisation of significantly differentially expressed genes 
```{r}
##==========================================================================================================
## 1.1 Genes significantly differentially expressed in clothianidin and imidacloprid group
##==========================================================================================================
### Organise data ####
## Select relevant gene counts from DESeq results (de_genes)
# normalise counts 
normalised_counts_df <- t(scale(t(txi_counts$counts), center = TRUE, scale = TRUE))

## Select significant gene counts
significant_de_counts <- normalised_counts_df[de_genes_CLO,]

## Convert significant counts into a dataframe:
significant_de_counts_df <- as.data.frame(significant_de_counts)

## Import txt file with gene descriptions
genes_for_heatmap <- read.table("./input/graphical_representation/de_genes_for_heatmap.txt", header = FALSE, sep = "\t", fill = TRUE)

## Double check this order matches with that in significant_de_counts_df 
genes_for_heatmap_LOC <- gsub(genes_for_heatmap$V1,  pattern = ".*[:(:]", replacement = "")
genes_for_heatmap_LOC <- gsub(genes_for_heatmap_LOC, pattern = "[:):]", replacement = "")

# Check order matches with rownames of heatmap object 
if (any(genes_for_heatmap_LOC != rownames(significant_de_counts_df))) {
    stop("genes not matched!") 
}

# Remove object 
rm(genes_for_heatmap_LOC)

## Add column with gene information:
significant_de_counts_df$gene_name <- rownames(significant_de_counts_df)

### Reorder based on logFC's 
significant_de_counts_df <- significant_de_counts_df[match(significant_de_counts_df$gene_name, rownames(de_genes_lfc)),]

# Add rownames with gene descriptions
rownames(significant_de_counts_df) <- genes_for_heatmap$V1

significant_de_counts_df$gene_name <- rownames(significant_de_counts_df)

## Reshape using 'melt':
significant_de_counts_df_melt <- melt(significant_de_counts_df)

## Rename column names
colnames(significant_de_counts_df_melt) <- c("gene_name", "sample", "normalised_counts")

# Add sample numbers to columns 
significant_de_counts_df_melt$sample  <- gsub(x = significant_de_counts_df_melt$sample, pattern = "-C.*", replacement = "")
significant_de_counts_df_melt$sample  <- paste(significant_de_counts_df_melt$sample, +rep(1:4,2))
  
  
```

# 1.2) Function to plot heatmap
```{r}

plot_heatmap <- function(input) {
                ggplot(input, aes(sample, reorder(gene_name, normalised_counts))) +
                geom_tile(aes(fill = normalised_counts)) +
                geom_vline(xintercept = c(4.5, 8.5)) +
                scale_fill_gradient2(low = "black", mid = "white", high = "red") +
                xlab("Samples") +
                ylab("") +
                theme(legend.title = element_text(size = 15, face = "bold"),
                      legend.text = element_text(size = 15, face = "bold"),
                      axis.text.x = element_text(angle = 45, hjust = 1, size = 8, face = "bold"),
                      axis.text.y = element_text(size = 8, face = "bold"),
                      axis.title = element_text(size = 14,face = "bold"),
                      panel.grid.major = element_blank(),
                      panel.grid.minor = element_blank(),
                      panel.background = element_blank()) +
                theme(legend.position = "top") +
                scale_y_discrete(position = "right") +
                geom_hline(yintercept = c(22.5, 23.5)) 
     
}


```

# 1.3) Plot heatmap and save 
```{r}
tiff("results/graphical_representation/de_genes_heatmap.tiff", width = 15, height = 7, units = 'in', res = 200)
plot_heatmap(significant_de_counts_df_melt) 
dev.off()
```



