---
title: "Bombus pesticide exposure: gene expression in head"
output: dexseq_analysis.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction
This script is for differential exon usage analysis of exon quantification data. Input files should be a count file (.fb.txt) generated by [HTSeq]("http://htseq.readthedocs.io/en/release_0.10.0/") for each sample and an annotation file (gff) generated by ```dexseq_prepare_annotation.py```, a python script provided with [DEXSeq]("https://bioconductor.org/packages/release/bioc/html/DEXSeq.html").

This script will import count files from HTSeq and use the R package DEXSeq to test for differential exon usage. This script produces a results table of genes that are significantly differentially expressed. This script uses the LRT test in DEXSeq to test for differential exon usage between treatments, which is the only option provided for analysis. Unlike DESeq2, there is no Wald test comparison. 

```{r, message = FALSE}
## Load libraries:
libraries <- c("DEXSeq", "ggplot2", "ggpubr", "DESeq2")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}
```

## Step One: Read in HTSeq count files:
We used HTSeq to extract read counts for each exon bin. 

```{r, message = FALSE}
## Read in files:
countfiles <- list.files("./results",
                        pattern = "sam.counts$",
                        full.names = TRUE)

## Read in GFF containing exon information:
flattenedfile <- list.files("./results",
                        pattern = "gff$",
                        full.names = TRUE)

## Extraction the name of treatments present within the data files:
countfiles_tmp <- gsub("-.*", "", countfiles)
treatments <- unique(gsub("./results/", "", countfiles_tmp))
control <- "CON"
non_control_treatments <- treatments[ !treatments %in% control]
```

## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r, message = FALSE}
## Suppress start up messages as recommended in DEXSeq walkthrough:
suppressPackageStartupMessages( library( "DEXSeq" ) )

## Read in batch information:
samples_information <- read.table(file = "./sample_information.txt",
                                  header = FALSE)

## Update header information:
colnames(samples_information) <- c("sample_name", "treatment", "room", "batch")
rownames(samples_information) <- samples_information$sample_name

## Ensure covariates are factors:
samples_information$batch <- as.factor(samples_information$batch)
samples_information$room <- as.factor(samples_information$room)
samples_information$sample_name <- NULL

## Use safer characters - DEXSeq complains about '-'
rownames(samples_information) <- gsub("-", "_", rownames(samples_information))

## Make sure 'CON' is the reference:
samples_information$treatment <- relevel(x = samples_information$treatment,
                                         ref = "CON")
```

## Step Three: Create a dexseq object for differential exon usage analysis
We create a dexseq object using:
- The count files (reads aligned per exon) generated by HTSeq.
- The sampleData file outlining the specific treatment each sample is assigned to.
- The variables compared within the GLM used DEXSeq.
- The "flattened" annotation file (gff) generated by the python script () provided with DEXSeq. 

```{r, message = FALSE}
## Create models for testing:
formulafullmodel <- ~sample + exon + batch:exon + room:exon + treatment:exon
formulareducedmodel <- ~sample + exon + batch:exon + room:exon

## For each pesticide treatment:
for (pesticide in non_control_treatments){
        ## Extract input files related to treatment of interest as well
        ## as control samples:
        countfiles_tmp <- c(grep(pesticide, countfiles, value = TRUE),
                           grep(control, countfiles, value = TRUE))
        print(countfiles_tmp)
        ## Extract sample information related to treatment of interest
        ## as well as control samples:
        sampleinfo_tmp <- rbind(subset(samples_information,
                                       treatment == pesticide),
                               subset(samples_information,
                                      treatment == control))
        print(sampleinfo_tmp)
        ## Create DEXSeq object:
        dexseq_object <- DEXSeqDataSetFromHTSeq(
                        countfiles_tmp,
                        sampleData = sampleinfo_tmp,
                        design = ~ sample + exon + treatment:exon,
                        flattenedfile = flattenedfile)
        ## Normalisation of exon counts by library sizes:
        dexseq_object <- estimateSizeFactors(dexseq_object)
        ## Estimate dispersion values:
        dexseq_object <- estimateDispersions(dexseq_object,
                                    formula = formulafullmodel)
        ## Test for differential exon usage between treatment groups:
        dexseq_object <- testForDEU(dexseq_object,
                           reducedModel = formulareducedmodel,
                           fullModel = formulafullmodel)
        ## Estimation of fold changes in exon usage between treatments:
        dexseq_object <- estimateExonFoldChanges(dexseq_object,
                                                fitExpToVar = "treatment")
        ## Subset results:
        dexseq_object_results <- DEXSeqResults(dexseq_object)
        dexseq_object_results_df <- as.data.frame(dexseq_object_results)
        ## Create a directory for output:
        output_directory <-  paste("results/", pesticide,
                                   "_differential_exon_raw_all/", sep = "")
        dir.create(path = output_directory,
                   recursive = TRUE)
        ## Write results to file:
        write.table(dexseq_object_results_df,
            file = paste(output_directory, "dexseq_object_results_all_",
                         pesticide, "_batch_room_corrected.txt",
                         sep = ""),
            row.names = FALSE,
            col.names = TRUE,
            sep = "\t",
            quote = FALSE)
        ## Subset columns of interest (groupID, pvalue and padj)
        dexseq_object_results_pvalues <- dexseq_object_results_df[, c(1, 6, 7)]
        write.table(dexseq_object_results_pvalues,
            file = paste(output_directory, "dexseq_object_results_all_",
                         pesticide, "_batch_room_corrected_pvalues.txt",
                         sep = ""),
            row.names = FALSE,
            col.names = TRUE,
            sep = "\t",
            quote = FALSE)
}
```

Additional code for visualisation of significant DEU genes:

```{r, message = FALSE}
# Plot MA-plot
plotMA(dexseq_object_results,
       cex = 0.8,
       alpha = 0.05,
       ylim = c(-5, 5))

## Sort data by significance for the visualisation of DE:
dexseq_object_results_sig <- dexseq_object_results_df[order(dexseq_object_results_df$padj), ]

# For examination of DEU for individual genes, use:
plotDEXSeq(dexseq_object_results,
           geneID = dexseq_object_results_sig$groupID[[1]],
           fitExpToVar = "treatment",
           norCounts = TRUE,
           displayTranscripts = TRUE,
           legend = TRUE,
           cex.axis = 1.2,
           cex = 1.3,
           lwd = 2)
```
