--- 
title: "Differential gene expression analyses"
output: 2018-03-13-DE.nb.html
--- 
---
Introduction:  
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by Kallisto (https://pachterlab.github.io/kallisto/manual) for each sample and .txt file of genome transcripts and their corresponding gene ids (LOC numbers)
This script will import abundance files from Kallisto and use the R package 'DESeq2' (http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed, following DESeq2 analysis. The results of which can then be used for further analysis. 
---

```{r}
## load packages 
library(tximport)
library(readr)
library(DESeq2)
library(ggplot2)
library(ggfortify)
library(ReportingTools)
library(vsn)
library(RColorBrewer)
library(readr)

## set wd 
setwd("/Users/isabelfletcher/Documents/LIDo/R1/RNAseq/2018-02-28-Differential_Expression")
```

## Step One: Input results from kallisto
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$data <- "." 
paths$kallisto_output <- file.path(paths$data, "2018-02-13-kallisto_quantification")

## Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", value = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

## Save path to use later
kallisto_quant_file_path <- "2018-01-11-kallisto_quantification"

## Automatically extract file names from kallisto output for colnames
names_tmp <- gsub(x = paths$kallisto_files, pattern = "./2018-01-11-kallisto_quantification/2016-Bter-", replacement = "")
names_tmp <- gsub(x = names_tmp, pattern = "-head.+", replacement = "")
names <- gsub(x = names_tmp, pattern = ".*Bter-", replacement = "")
rm(names_tmp)
```

## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport
```{r}
## Extract sample names and put into df for tximport
samples_tmp <- gsub(x = paths$kallisto_files, pattern = ".*Bter-", replacement = "")
samples <- gsub(x = samples_tmp,     pattern = "-head.*",     replacement = "")
samples  <- data.frame(treatment = samples)
rownames(samples) <- samples$treatment
samples$treatment <- gsub(x = samples$treatment ,     pattern = "-.*",     replacement = "")
rm(samples_tmp)
```

## Step Three: Import expression values using Tximport
Tximport reads in expression values generated by kallisto for transcripts.  
These values are summarised per gene using tximport.  

```{r}
## Read in file corresponding to transcripts to gene ids
tx2gene <- read.table("./rna_and_corresponding_gene_ids.txt", header = F)
head(tx2gene)

# Assign column names
# Column 1: transcript name
# Column 2: gene name
colnames(tx2gene) <- c("TXNAME","GENEID")

## Use tximport on kallisto files 
## Counts are output for TPM values scaled to gene length:
txi <- tximport(paths$kallisto_files, type = "kallisto", 
                tx2gene = tx2gene,
                countsFromAbundance = c("lengthScaledTPM"))

## Convert to dataframe for potential plots:
colnames(txi$counts) <- names #label columns with treatments
## Extract count values:
txi_counts_df <- as.data.frame(txi$counts)
## Save extracted count values:
save(txi_counts_df, file = "txi_counts_df.Rdata")

# Save object for use in other scripts
save(txi, file = "txi.Rdata")

```
## Step Four: Run DESeq2 for differential expression analysis:
##================================================================================================
## 1) Which genes are differentially expressed across all three treatments?
##================================================================================================
## Running a likelihood ratio test 
# Make sure that all files for each condition of interest are read into the DESEq object
```{r}
## Construct DESeq data set from txi object and sample information
deseq_txi <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ treatment) 

## Relevel treatments, so CON is defined as base
deseq_txi$treatment <- relevel(deseq_txi$treatment, "CON")
design(deseq_txi) <- ~ treatment

## 1) Running a likelihood ratio test - for DE comparison across all treatments
# Make sure that all files for each condition of interest are read into the DESEq object:
deseq_object_LRT <- DESeq(deseq_txi, test = "LRT", reduced = ~1)

## Generate a report of differentially expressed genes and output in folder 'reports':
deseq_report <- HTMLReport(shortName = 'RNAseq_analysis_with_DESeq2',
                         title = 'RNA-seq analysis of differential expression using DESeq2',
                         reportDirectory = "./reports")

## Publish report:
publish(deseq_object_LRT,deseq_report, pvalueCutoff = 0.05,
        factor = colData(deseq_object_LRT)$treatment,
        reportDir = "./reports")
        
finish(deseq_report)

##================================================================================================
## Subset the fold change differences between the treatments:
##================================================================================================
## Generate results table for DE genes between treatments, specifying contrasts:
## Comparison of clothianidin vs. control:
deseq_results_CON_CLO <- results(deseq_object_LRT, contrast = c("treatment", "CON", "CLO"))

## Comparison of imidacloprid vs. control:
deseq_results_CON_IMI <- results(deseq_object_LRT, contrast = c("treatment", "CON", "IMI"))

## Log fold change shrinkage for ranking and visualisation, provide lfcshrink() with the DESeq object
deseq_results_shrink_CON_CLO <- lfcShrink(deseq_object_LRT, contrast = c("treatment", "CON", "CLO"))
deseq_results_shrink_CON_IMI <- lfcShrink(deseq_object_LRT, contrast = c("treatment", "CON", "IMI"))

# Order results by adjusted p-value, which will be the same for each treatment comparison but log fold changes will occur:
deseq_results_ordered_CON_CLO <- deseq_results_shrink_CON_CLO[order(deseq_results_shrink_CON_CLO$padj),]
deseq_results_ordered_CON_IMI <- deseq_results_shrink_CON_IMI[order(deseq_results_shrink_CON_IMI$padj),]

## Create table of significant genes
deseq_results_significant <- deseq_results_ordered_CON_CLO[which(deseq_results_ordered_CON_CLO$padj < 0.05), ]
nrow(deseq_results_significant) # 15

# Save table for further use
save(deseq_results_significant, file = "deseq_results_significant.Rdata")
```
## Step Five: Run DESeq2 (Wald test) for differential expression analysis:
##================================================================================================
## Which genes are differentially expressed between each treatment and control?
##================================================================================================
## Running a Wald t-test (default test within DESeq2)

```{r}
## 2) Pairwise comparisons of treatments and control groups
deseq_object_pairwise <- DESeq(deseq_txi)

# Save objects for later use
save(deseq_object_pairwise, file = "deseq_object_pairwise.Rdata")
save(deseq_object_LRT, file = "deseq_object_LRT.Rdata")

# 2) Pairwise comparisons between treatments 
# Generate results table for DE genes, specifying pairwise contrasts
deseq_results_CON_CLO <- results(deseq_object_pairwise, contrast = c("treatment", "CON", "CLO"))
deseq_results_CON_IMI <- results(deseq_object_pairwise, contrast = c("treatment", "CON", "IMI"))

## Log fold change shrinkage for ranking and visualisation, provide lfcshrink() with the DESeq object 
deseq_results_CON_CLO_shrink <- lfcShrink(deseq_object_pairwise, contrast = c("treatment", "CON","CLO"))
deseq_results_CON_IMI_shrink <- lfcShrink(deseq_object_pairwise, contrast = c("treatment", "CON","IMI"))

# Order results by p value
deseq_results_CON_CLO_ordered <- deseq_results_CON_CLO_shrink[order(deseq_results_CON_CLO_shrink$padj),]
deseq_results_CON_IMI_ordered <- deseq_results_CON_IMI_shrink[order(deseq_results_CON_IMI_shrink$padj),]

## Create tables of significant genes 
deseq_results_CON_CLO_significant <- deseq_results_CON_CLO_ordered[which(deseq_results_CON_CLO_ordered$padj < 0.05), ]
deseq_results_CON_IMI_significant <- deseq_results_CON_IMI_ordered[which(deseq_results_CON_IMI_ordered$padj < 0.05), ]

# Save tables for further use
save(deseq_results_CON_CLO_significant, file = "deseq_results_CON_CLO_significant.Rdata")
save(deseq_results_CON_IMI_significant, file = "deseq_results_CON_IMI_significant.Rdata")

# Put significant DE gene names into an object
de_genes_CON_CLO <- rownames(deseq_results_CON_CLO_significant)
de_genes_CON_IMI <- rownames(deseq_results_CON_IMI_significant)

de_genes_CON_CLO <- gsub(x = de_genes_CON_CLO, pattern = ".+rna_", replacement = "")
de_genes_CON_IMI <- gsub(x = de_genes_CON_IMI, pattern = "[:.:][0-9].+", replacement = "")

# Export objectS into .txt file, for GO analysis
write(de_genes_CON_CLO, "de_genes_CON_CLO.txt")
save(de_genes_CON_CLO, file = "de_genes_CON_CLO.Rdata")

write(de_genes_CON_IMI, "de_genes_CON_IMI.txt")
save(de_genes_CON_IMI, file = "de_genes_CON_IMI.Rdata")

```
# Explore genes that have largest log-fold changes
```{r}
# Order results table by log fold change
deseq_results_CON_CLO_significant_lfc <- deseq_results_CON_CLO_significant[order(deseq_results_CON_CLO_significant$log2FoldChange),]

# Genes which are more highly expressed in CLO group
deseq_results_CON_CLO_significant_high <- deseq_results_CON_CLO_significant_lfc[which(deseq_results_CON_CLO_significant_lfc$log2FoldChange < 0),]

# Genes which are more highly expressed in CON group
deseq_results_CON_CLO_significant_low <- deseq_results_CON_CLO_significant_lfc[which(deseq_results_CON_CLO_significant_lfc$log2FoldChange > 0),]

```

## Step Six: Principal Components Analysis
```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object_LRT, blind = FALSE)
rld <- rlog(deseq_object_LRT, blind = FALSE)

# this gives log2(n + 1)
ntd        <- normTransform(deseq_object_LRT)

## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))

## Extract first and second principal components:
pca_data    <- plotPCA(vsd, intgroup = c("treatment"), returnData = TRUE)

## Extract percentages for first two principal components:
percentVar <- round(100 * attr(pca_data, "percentVar"))

## Plot PCA:
ggplot(pca_data, aes(PC1, PC2, color = treatment, shape = treatment)) +
       geom_point(size = 5, shape = 16) +
       xlab(paste0("PC1: ",percentVar[1],"% variance")) +
       ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
       coord_fixed() +
       scale_color_manual(values = c("blue", "grey", "black")) +
       theme_bw() +
       theme(axis.title = element_text(family = "Arial", color = "#666666", face = "bold", size = 16)) 

        
```



