--- 
title: "Differential gene expression analyses"
output: 2018-03-13-DE.nb.html
--- 
---
Introduction:  
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by Kallisto (https://pachterlab.github.io/kallisto/manual) for each sample and .txt file of genome transcripts and their corresponding gene ids (LOC numbers)
This script will import abundance files from Kallisto and use the R package 'DESeq2' (http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed, following DESeq2 analysis. The results of which can then be used for further analysis. 
---

```{r}
## load packages: 
library(tximport)
library(readr)
library(DESeq2)
library(ggplot2)
library(gplots)
library(ggfortify)
library(ReportingTools)
library(vsn)
library(RColorBrewer)
library(readr)

## set wd 
setwd("/Volumes/apocrita/autoScratch/2017-08-11_bombus_transcriptomics_pesticides_exp_1/results/2018-02-24_multi_analysis/2018-03-05_ifletcher_analysis/2018-02-28-Differential_Expression")
```

### Step One: Input results from kallisto
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$data <- "." 
paths$kallisto_output <- file.path(paths$data, "2018-02-13-kallisto_quantification")

## Set relative paths:
paths$kallisto_files_relative <- grep(x = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", value = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

## Save path to use later
kallisto_quant_file_path <- "2018-01-11-kallisto_quantification"

## Automatically extract file names from kallisto output for colnames
names_tmp <- gsub(x = paths$kallisto_files, pattern = "./2018-01-11-kallisto_quantification/2016-Bter-", replacement = "")
names_tmp <- gsub(x = names_tmp, pattern = "-head.+", replacement = "")
names <- gsub(x = names_tmp, pattern = ".*Bter-", replacement = "")
rm(names_tmp)
```

### Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport
```{r}
## Extract sample names and put into df for tximport
samples_tmp <- gsub(x = paths$kallisto_files, pattern = ".*Bter-", replacement = "")
samples <- gsub(x = samples_tmp,     pattern = "-head.*",     replacement = "")
samples  <- data.frame(treatment = samples)
rownames(samples) <- samples$treatment
samples$treatment <- gsub(x = samples$treatment ,     pattern = "-.*",     replacement = "")
rm(samples_tmp)
```

### Step Three: Import expression values using Tximport
Tximport reads in expression values generated by kallisto for transcripts.  
These values are summarised per gene using tximport.  

```{r}
## Read in file corresponding to transcripts to gene ids
tx2gene <- read.table("./rna_and_corresponding_gene_ids.txt", header = F)

# Assign column names
# Column 1: transcript name
# Column 2: gene name
colnames(tx2gene) <- c("TXNAME","GENEID")

## Use tximport on kallisto files 
## Counts are output for TPM values scaled to gene length:
txi_transcripts <- tximport(paths$kallisto_files, type = "kallisto", txOut = T)

## Convert to dataframe for potential plots:
colnames(txi_transcripts$counts) <- names #label columns with treatments
## Extract count values:
txi_transcripts_counts_df <- as.data.frame(txi_transcripts$counts)

## For transcripts, rename column names:
rownames(txi_transcripts_counts_df) <- gsub(x = rownames(txi_transcripts_counts_df), pattern = ".+rna_", replacement = "")
head(txi_transcripts_counts_df)

## Save extracted count values:
save(txi_transcripts_counts_df, file = "txi_transcripts_counts_df.Rdata")

# Save object for use in other scripts
save(txi_transcripts, file = "txi_transcripts.Rdata")

```
### Step Four: Run DESeq2 for differential expression analysis:

### 1) Which transcripts are differentially expressed across all three treatments?

### Running a likelihood ratio test 
Make sure that all files for each condition of interest are read into the DESEq object
```{r}
## Construct DESeq data set from txi object and sample information
deseq_txi_transcripts <- DESeqDataSetFromTximport(txi_transcripts, 
                                      colData = samples, 
                                      design = ~ treatment) 

## Relevel treatments, so CON is defined as base
deseq_txi_transcripts$treatment <- relevel(deseq_txi_transcripts$treatment, "CON")
design(deseq_txi_transcripts) <- ~ treatment

## 1) Running a likelihood ratio test - for DE comparison across all treatments
# Make sure that all files for each condition of interest are read into the DESEq object:
deseq_trans_object_LRT <- DESeq(deseq_txi_transcripts, test = "LRT", reduced = ~ 1)
rownames(deseq_trans_object_LRT) <- gsub(x = rownames(deseq_trans_object_LRT), pattern = ".+rna_", replacement = "")

## Generate a report of differentially expressed genes and output in folder 'reports':
deseq_trans_report <- HTMLReport(shortName = 'Transcript_analysis_with_DESeq2',
                         title = 'Transcript analysis of differential expressed transcripts using DESeq2',
                         reportDirectory = "./trans_reports")

## Publish report:
publish(deseq_trans_object_LRT,deseq_trans_report, pvalueCutoff = 0.05,
        factor = colData(deseq_trans_object_LRT)$treatment,
        reportDir = "./trans_reports")
        
finish(deseq_trans_report)

##===========================================================
## Subset the fold change differences between the treatments:
##===========================================================
## Generate results table for DE genes between treatments, specifying contrasts:
## Comparison of clothianidin vs. control:

# Set colour pattern for plotting
scalepurpleyellow <- colorRampPalette(colors = c("blue", "purple", "yellow"))(250)

## The function takes a z-normalised dataframe as input
plot_heatmap3 <- function(input){
                  heatmap.2(input, 
                  col = scalepurpleyellow, 
                  trace = "none",
                  margins = c(10,14), 
                  sepwidth = c(0.01, 0.01), 
                  sepcolor = "black", 
                  colsep = 1:ncol(input),
                  rowsep = 1:nrow(input),
                  cexRow = 1,
                  Colv = NULL,
                  labRow = rownames(input))
}

## Create function to generate results tables for pairwise comparisons:
generate_pairwise_heatmap <- function(deseq_input, treatment_1, treatment_2, counts_input){
                          ## Shrink log fold changes:
                          deseq_results_shrink <- lfcShrink(deseq_input, contrast = c("treatment", treatment_1, treatment_2))
                          ## Reorder results:
                          deseq_results_shrink_ordered <- deseq_results_shrink[order(deseq_results_shrink$padj),]
                          ## Subset significant values:
                          de_genes <- deseq_results_shrink_ordered[which(deseq_results_shrink_ordered$padj < 0.05), ]
                          ## Print to console:
                          print(nrow(de_genes))
                          ## Select relevant gene counts from DESeq results (de_genes)
                          # normalise counts 
                          normalised_counts_df <- t(scale(t(counts_input), center = TRUE, scale = TRUE))
                          # Select significant gene counts
                          significant_de_counts <- normalised_counts_df[row.names(de_genes),]
                          ## Return:
                          return(plot_heatmap3(significant_de_counts))
}

## Call function:
generate_pairwise_heatmap(deseq_trans_object_LRT, "CON", "CLO", txi_transcripts_counts_df) 

## Export results
de_transcripts_res <- results(deseq_trans_object_LRT, contrast = c("treatment", "CON", "CLO"))

## Subset statistically significant sites:
de_transcripts_res_significant <- de_transcripts_res[which(de_transcripts_res$padj < 0.05), ]

## Update the transcript names to extract gene IDs:
tx2gene$TXNAME <- gsub(x = tx2gene$TXNAME, pattern = ".+rna_", replacement = "")

## Equate to how many genes?
tx2gene_subset_trans <- tx2gene[tx2gene$TXNAME %in% rownames(de_transcripts_res_significant),]

## Extract the number of unique genes:
tx2gene_trans_unique_genes <- unique(tx2gene_subset_trans$GENEID)
tx2gene_trans_unique_genes <- as.character(unlist(tx2gene_trans_unique_genes))

## Subset differentially expressed transcripts with differentially expressed genes:
tx2gene_trans_unique_genes[tx2gene_trans_unique_genes %in% de_unique_genes]

# Save table for further use
#save(deseq_results_significant, file = "deseq_results_significant.Rdata")
```
### Step Five: Run DESeq2 (Wald test) for differential expression analysis:

### Which genes are differentially expressed between each treatment and control?
Running a Wald t-test (default test within DESeq2)

```{r}
## 2) Pairwise comparisons of treatments and control groups
deseq_object_pairwise <- DESeq(deseq_txi)

## Update rowname for pairwise comparison:
rownames(deseq_object_pairwise) <- txi_counts_df_rownames
rownames(txi_counts_df) <- txi_counts_df_rownames

res <- results(deseq_object_pairwise, contrast=c("treatment","CON","IMI"))
de_genes_res <- res[which(res$padj < 0.05), ]
genes_to_subset <- rownames(de_genes_res)

## Update the transcript names to extract gene IDs:
tx2gene$TXNAME <- gsub(x = tx2gene$TXNAME, pattern = ".+rna_", replacement = "")

tx2gene_subset <- tx2gene[tx2gene$TXNAME %in% genes_to_subset,]
length(unique(tx2gene_subset$TXNAME))
length(unique(tx2gene_subset$GENEID))

# Save objects for later use
save(deseq_object_pairwise, file = "deseq_object_pairwise.Rdata")
save(deseq_object_LRT, file = "deseq_object_LRT.Rdata")

# 2) Pairwise comparisons between treatments 
generate_pairwise_heatmap(deseq_object_pairwise, "CON", "CLO", txi_counts_df) 
generate_pairwise_heatmap(deseq_object_pairwise, "CON", "IMI", txi_counts_df) 

# Save tables for further use
save(deseq_results_CON_CLO_significant, file = "deseq_results_CON_CLO_significant.Rdata")
save(deseq_results_CON_IMI_significant, file = "deseq_results_CON_IMI_significant.Rdata")

# Put significant DE gene names into an object
de_genes <- rownames(deseq_results_CON_CLO)

## Rename rownames:
txi_counts_df_rownames<- rownames(txi_counts_df)
txi_counts_df_rownames <- gsub(x = txi_counts_df_rownames, pattern = ".+rna_", replacement = "")
#txi_counts_df_rownames <- gsub(x = txi_counts_df_rownames, pattern = "[:.:][0-9].+", replacement = "")

## Rename rownames:
rownames(txi_counts_df) <- txi_counts_df_rownames

## Subset rows of interest:
generate_pairwise_heatmap(deseq_object_pairwise, "CON", "CLO", txi_counts_df) 



generate_pairwise_heatmap(deseq_object_pairwise, "CON", "CLO", de_genes_transcripts) 


# Export objectS into .txt file, for GO analysis
write(de_genes_CON_CLO, "de_genes_CON_CLO.txt")
save(de_genes_CON_CLO, file = "de_genes_CON_CLO.Rdata")

write(de_genes_CON_IMI, "de_genes_CON_IMI.txt")
save(de_genes_CON_IMI, file = "de_genes_CON_IMI.Rdata")

```
### Explore genes that have largest log-fold changes
```{r}
# Order results table by log fold change
deseq_results_CON_CLO_significant_lfc <- deseq_results_CON_CLO_significant[order(deseq_results_CON_CLO_significant$log2FoldChange),]

# Genes which are more highly expressed in CLO group
deseq_results_CON_CLO_significant_high <- deseq_results_CON_CLO_significant_lfc[which(deseq_results_CON_CLO_significant_lfc$log2FoldChange < 0),]

# Genes which are more highly expressed in CON group
deseq_results_CON_CLO_significant_low <- deseq_results_CON_CLO_significant_lfc[which(deseq_results_CON_CLO_significant_lfc$log2FoldChange > 0),]

```

### Step Six: Principal Components Analysis

```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object_LRT, blind = FALSE)
rld <- rlog(deseq_object_LRT, blind = FALSE)

# this gives log2(n + 1)
ntd        <- normTransform(deseq_object_LRT)

## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))

## Extract first and second principal components:
pca_data    <- plotPCA(vsd, intgroup = c("treatment"), returnData = TRUE)

## Extract percentages for first two principal components:
percentVar <- round(100 * attr(pca_data, "percentVar"))

## Plot PCA:
ggplot(pca_data, aes(PC1, PC2, color = treatment, shape = treatment)) +
       geom_point(size = 5, shape = 16) +
       xlab(paste0("PC1: ",percentVar[1],"% variance")) +
       ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
       coord_fixed() +
       scale_color_manual(values = c("blue", "grey", "black")) +
       theme_bw() +
       theme(axis.title = element_text(family = "Arial", color = "#666666", face = "bold", size = 16)) 
    
```



