--- 
title: "Bombus pesticide exposure: gene expression in head"
output: gene_level_3treatments.html
--- 

# Introduction
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by Kallisto
(https://pachterlab.github.io/kallisto/manual) for each sample and .txt file of genome transcripts 
and their corresponding gene ids (LOC numbers)/.

This script will import abundance files from Kallisto and use the R package 'DESeq2'
(http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for 
differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed. 
This can then be used for further analysis. 


```{r}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2", "ggplot2", "gplots", 
               "ggfortify", "ReportingTools", "vsn", "RColorBrewer")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
```


## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.

```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$kallisto_output <- "2018-02-13-kallisto_quantification"

# Set relative paths:
paths$kallisto_files_relative <- grep(x       = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", 
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
names_tmp <- paths$kallisto_files
names_tmp <- gsub(x = names_tmp, pattern     = "./2018-01-11-kallisto_quantification/2016-Bter-", 
                  replacement = "")
names_tmp <- gsub(x = names_tmp, pattern = "-head.+", replacement = "")
names     <- gsub(x = names_tmp, pattern = ".*Bter-", replacement = "")
rm(names_tmp)
```


## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples_tmp <- gsub(x = paths$kallisto_files, pattern = ".*Bter-", replacement = "")
samples     <- gsub(x = samples_tmp,          pattern = "-head.*",replacement = "")
samples     <- data.frame(treatment = samples)
rownames(samples) <- samples$treatment
samples$treatment <- gsub(x = samples$treatment ,     pattern = "-.*",     replacement = "")
rm(samples_tmp)
```


## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
tx2gene <- read.table("./rna_and_corresponding_gene_ids.txt", header = FALSE)

# Assign column names
# Column 1: transcript name
# Column 2: gene name
colnames(tx2gene) <- c("TXNAME","GENEID")

# Use tximport on kallisto files 
## Counts are output for TPM values scaled to gene length:
txi <- tximport(paths$kallisto_files, 
                type    = "kallisto", 
                tx2gene = tx2gene)

## Convert to dataframe for potential plots:
colnames(txi$counts) <- names #label columns with treatments
## Extract count values:
txi_counts_df <- as.data.frame(txi$counts)
## Save extracted count values:
save(txi_counts_df, file = "txi_counts_df.Rdata")

# Save object for use in other scripts
save(txi, file = "txi.Rdata")
```

### Step Four: Run DESeq2 for differential expression analysis:

For identifying differentially expressed genes, a DESeq object is generated containing:
    a. sample names
    b. treatments
    
Within the present analysis, three types of questions can be immediately asked:
    1. What genes are differentially expressed across all three treatments?
    2. What genes are differentially expressed within clothianidin-exposed workers in comparison to control?
    3. What genes are differentially expressed within imidacloprid-exposed workers in comparison to control?

```{r}
## Construct DESeq data set from txi object and sample information
deseq_txi <- DESeqDataSetFromTximport(txi, 
                                      colData = samples, 
                                      design = ~ treatment) 

## Relevel treatments, so CON is defined as base
deseq_txi$treatment <- relevel(deseq_txi$treatment, "CON")
design(deseq_txi)   <- ~ treatment

##=========================================================================================
## 4.1) Which genes are differentially expressed across all three treatments?
##=========================================================================================

## 1) Running a likelihood ratio test - for DE comparison across all treatments
## Make sure that all files for each condition of interest are read into the DESEq object:
deseq_object_LRT <- DESeq(deseq_txi, test = "LRT", reduced = ~ 1)

##=========================================================================================
## 4.2) Identify log fold changes between treatments for genes differentially expressed:
## a) Clothianidin vs. control
##=========================================================================================

## The LRT will output a single adjusted p-value for each significant gene.
## By default, the log fold change reported will be between the first two treatments, which
## 'control' and 'clothianidin' but you can also specify treatments to compare using the
## 'contrast' parameter:
## Generate results table for DE genes between treatments, specifying contrasts:
deseq_results_CON_CLO <- results(deseq_object_LRT, contrast = c("treatment", "CON", "CLO"))

test_CLO <- results(deseq_object_LRT, contrast = c("treatment", "CON", "CLO"), test = "Wald") 

## Log fold change shrinkage for ranking and visualisation, provide lfcshrink() with the DESeq object
deseq_results_shrink_CON_CLO <- lfcShrink(deseq_object_LRT, contrast = c("treatment", "CON", "CLO"))

## Order results by adjusted p-value, which will be the same for each treatment comparison but log fold changes will occur:
deseq_results_ordered_CON_CLO    <- deseq_results_shrink_CON_CLO[order(deseq_results_shrink_CON_CLO$padj),]

## Subset significant genes:
deseq_results_significant_CON_CLO <- deseq_results_ordered_CON_CLO[which(deseq_results_ordered_CON_CLO$padj < 0.05), ]

## To identify direction of expression (over or under expressed):
# Order DE genes by lfc
de_logchange_CON_CLO <- deseq_results_significant_CON_CLO[order(deseq_results_significant_CON_CLO$log2FoldChange),]

# Genes more highly expressed in CLO treatment, compared to CON i.e. negative lfc values
de_logchange_CON_CLO_high <- de_logchange_CON_CLO[deseq_results_significant_CON_CLO$log2FoldChange < 0,]

# Genes less expressed in CLO treatment, compared to CON
de_logchange_CON_CLO_low <- de_logchange_CON_CLO[deseq_results_significant_CON_CLO$log2FoldChange > 0,]

# Save table for further use
save(deseq_results_significant, file = "deseq_results_significant.Rdata")

# Put significant DE gene names into an object
de_genes <- rownames(deseq_results_significant)

# Export object into .txt file, for GO analysis
write(de_genes, "de_genes.txt")
save(de_genes, file = "de_genes.Rdata")

##=========================================================================================
## 4.2) Identify log fold changes between treatments for genes differentially expressed:
## b) Imidacloprid vs. control
##=========================================================================================

## Control vs. Imidacloprid:
deseq_results_CON_IMI <- results(deseq_object_LRT, contrast = c("treatment", "CON", "IMI"))

test_IMI <- results(deseq_object_LRT, contrast = c("treatment", "CON", "IMI"), test = "Wald")

## Log fold change shrinkage for ranking and visualisation, provide lfcshrink() with the DESeq object
deseq_results_shrink_CON_IMI <- lfcShrink(deseq_object_LRT, contrast = c("treatment", "CON", "IMI"))

# Order results by adjusted p-value, which will be the same for each treatment comparison but log fold changes will occur:
deseq_results_ordered_CON_IMI <- deseq_results_shrink_CON_IMI[order(deseq_results_shrink_CON_IMI$padj),]

## Create table of significant genes
deseq_results_significant_CON_IMI <- deseq_results_ordered_CON_IMI[which(deseq_results_ordered_CON_CLO$padj < 0.05), ]

# Order DE genes by lfc 
de_logchange_CON_IMI <- deseq_results_significant_CON_IMI[order(deseq_results_significant_CON_IMI$log2FoldChange),]

# Genes more highly expressed in IMI treatment, compared to CON i.e. negative lfc values
de_logchange_CON_IMI_high <- deseq_results_significant_CON_IMI[deseq_results_significant_CON_IMI$log2FoldChange < 0,]

# Genes less expressed in IMI treatment, compared to CON i.e. positive lfc values
de_logchange_CON_IMI_low <- deseq_results_significant_CON_IMI[deseq_results_significant_CON_IMI$log2FoldChange > 0,]

##=========================================================================================
## 4.3 Generate a report on fold changes between treatements
##=========================================================================================

## Generate a report of differentially expressed genes and output in folder 'reports'.
## Report results in the generation of boxplots for each statistically significant gene:
deseq_report <- HTMLReport(shortName = 'RNAseq_analysis_with_DESeq2',
                         title = 'RNA-seq analysis of differential expressed transcripts using DESeq2',
                         reportDirectory = "./reports")

## Publish report:
publish(deseq_object_LRT,deseq_report, pvalueCutoff = 0.05,
        factor = colData(deseq_object_LRT)$treatment,
        reportDir = "./reports")

## Complete the report:        
finish(deseq_report)

##=========================================================================================
## 4.4 Visualise fold change differences between the treatments:
##=========================================================================================
## Generate results table for DE genes between treatments:

# Set colour pattern for plotting
scalepurpleyellow <- colorRampPalette(colors = c("blue", "purple", "yellow"))(250)

## The function takes a z-normalised dataframe as input
plot_heatmap3 <- function(input){
                  heatmap.2(input, 
                  col = scalepurpleyellow, 
                  trace = "none",
                  margins = c(10,14), 
                  sepwidth = c(0.01, 0.01), 
                  sepcolor = "black", 
                  colsep = 1:ncol(input),
                  rowsep = 1:nrow(input),
                  cexRow = 1,
                  Colv = NULL,
                  labRow = rownames(input))
}

## Create function to generate results tables for pairwise comparisons:
generate_pairwise_heatmap <- function(deseq_input, treatment_1, treatment_2, counts_input){
                          ## Shrink log fold changes:
                          deseq_results_shrink <- lfcShrink(deseq_input, contrast = c("treatment", treatment_1, treatment_2))
                          ## Reorder results:
                          deseq_results_shrink_ordered <- deseq_results_shrink[order(deseq_results_shrink$padj),]
                          ## Subset significant values:
                          de_genes <- deseq_results_shrink_ordered[which(deseq_results_shrink_ordered$padj < 0.05), ]
                          ## Print to console:
                          print(nrow(de_genes))
                          # normalise counts 
                          normalised_counts_df <- t(scale(t(counts_input), center = TRUE, scale = TRUE))
                          # Select significant gene counts
                          significant_de_counts <- normalised_counts_df[row.names(de_genes),]
                          ## Return:
                          return(plot_heatmap3(significant_de_counts))
}

## Call function:
generate_pairwise_heatmap(deseq_object_LRT, "CON", "CLO", txi_counts_df) 

```
### Step Five: Run DESeq2 (Wald test) for differential expression analysis:

### Which genes are differentially expressed between each treatment and control?
Running a Wald t-test (default test within DESeq2)

```{r}
##=========================================================================================================
## 5.1) Which genes are differentially expressed within clothianidin-exposed workers in comparison to control?
##=========================================================================================================
## Perform a pairwise comparison between treatments:
deseq_object_pairwise  <- DESeq(deseq_txi, test = "Wald") #fit model to txi object 

## Extract results for clothianidin vs. control:
deseq_results_pairwise_CLO <- results(deseq_object_pairwise, contrast = c("treatment", "CON", "CLO")) #results of model

## Save object for later use
save(deseq_object_pairwise, file = "deseq_object_pairwise.Rdata")

## lfcShrink(), which performs log2 fold change shrinkage for visualization and ranking of genes:
deseq_results_pairwise_CLO_shrink <- lfcShrink(dds = deseq_object_pairwise, contrast = c("treatment", "CON", "CLO"), res = deseq_results_pairwise_CLO)

# Results table ordered by smallest p value
deseq_results_pairwise_CLO_ordered <- deseq_results_pairwise_CLO_shrink[order(deseq_results_pairwise_CLO_shrink$pvalue),]

# No. genes p < 0.05
deseq_results_pairwise_CLO_significant <- deseq_results_pairwise_CLO_ordered[which(deseq_results_pairwise_CLO_ordered$padj < 0.05), ]

## Log fold changes in expression between clothianidin-exposed workers and control ###
# Genes which are more highly expressed in CLO group, compared to control 
deseq_pairwise_CLO_lfc_high <- deseq_results_pairwise_CLO_significant[which(deseq_results_pairwise_CLO_significant$log2FoldChange < 0),] # n = 21

# Genes which are less expressed in CLO group, compared to control
deseq_pairwise_CLO_lfc_low <- deseq_results_pairwise_CLO_significant[which(deseq_results_pairwise_CLO_significant$log2FoldChange > 0),] # n = 12

# Save results object for use in heatmap
de_genes_pairwise_CLO <- rownames(deseq_results_pairwise_CLO_significant)
save(de_genes_pairwise_CLO, file = "de_genes_pairwise_CLO.Rdata")

# Export gene names object into .txt file, for GO analysis genes of interest 
write(de_genes_pairwise_CLO, "de_genes_pairwise_CLO.txt")

##=========================================================================================================
## 5.2 Which genes are differentially expressed within imidacloprid-exposed workers in comparison to control?
##=========================================================================================================

## Perform a pairwise comparison between treatments
deseq_results_pairwise_IMI <- results(deseq_object_pairwise, contrast = c("treatment", "CON", "IMI")) #results of model

#lfcShrink(), which performs log2 fold change shrinkage for visualization and ranking of genes:
deseq_results_pairwise_IMI_shrink <- lfcShrink(dds = deseq_object_pairwise, contrast = c("treatment", "CON", "IMI"), res = deseq_results_pairwise_IMI)

# Results table ordered by smallest p value
deseq_results_pairwise_IMI_ordered <- deseq_results_pairwise_IMI_shrink[order(deseq_results_pairwise_IMI_shrink$pvalue),]

# No. genes p < 0.05
deseq_results_pairwise_IMI_significant <- deseq_results_pairwise_IMI_ordered[which(deseq_results_pairwise_IMI_ordered$padj < 0.05), ] # n = 1

```

### Step Six: Principal Components Analysis

```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object_LRT, blind = FALSE)
rld <- rlog(deseq_object_LRT, blind = FALSE)

# this gives log2(n + 1)
ntd        <- normTransform(deseq_object_LRT)

## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))

## Extract first and second principal components:
pca_data    <- plotPCA(vsd, intgroup = c("treatment"), returnData = TRUE)

## Extract percentages for first two principal components:
percentVar <- round(100 * attr(pca_data, "percentVar"))

## Plot PCA:
ggplot(pca_data, aes(PC1, PC2, color = treatment, shape = treatment)) +
       geom_point(size = 5, shape = 16) +
       xlab(paste0("PC1: ",percentVar[1],"% variance")) +
       ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
       coord_fixed() +
       scale_color_manual(values = c("blue", "grey", "black")) +
       theme_bw() +
       theme(axis.title = element_text(family = "Arial", color = "#666666", face = "bold", size = 16)) 
    
```



