---
title: "R Notebook"
output: html_notebook
---
# Introduction
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by [Kallisto](https://pachterlab.github.io/kallisto/manual) 
for each sample and .txt file of genome transcripts 
and their corresponding gene ids (LOC numbers)/.

This script will import abundance files from Kallisto and use the R package [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for 
differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed. 
This script uses the 'Wald' test in DESeq2 to test for differential expression across specified treatments.
This can then be used for further analysis. 

```{r}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2", "ggplot2", "gplots", "rhdf5",
               "plyr", "ggfortify", "ReportingTools", "vsn", "RColorBrewer")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
dir.create("results")
```

## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$kallisto_output <- "input/kallisto_quantification"

# Set relative paths:
paths$kallisto_files_relative <- grep(x       = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".tsv", 
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
filenames_tmp <- paths$kallisto_files
filenames_tmp <- gsub(x = filenames_tmp, pattern     = file.path(paths$kallisto_output, "2016-Bter-"), 
                  replacement = "")
filenames_tmp <- gsub(x = filenames_tmp, pattern = "-head.+", replacement = "")
filenames     <- gsub(x = filenames_tmp, pattern = ".*Bter-", replacement = "")
names(paths$kallisto_files) <- filenames 

for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
rm(filenames_tmp, filenames)
```


## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples_tmp <- gsub(x = paths$kallisto_files, pattern = ".*Bter-", replacement = "")
samples     <- gsub(x = samples_tmp,          pattern = "-head.*", replacement = "")
samples     <- data.frame(treatment = samples)
rownames(samples) <- samples$treatment
samples$treatment <- gsub(x = samples$treatment ,     pattern = "-.*",     replacement = "")
samples$treatment <- as.factor(samples$treatment)
samples$treatment <- relevel(x = samples$treatment, ref = "CON") 
rm(samples_tmp)
```


## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
tx2gene <- read.table(file      = "input/rna_and_corresponding_gene_ids.txt", 
                      header    = FALSE, 
                      col.names = c("TXNAME","GENEID")) # i.e., transcript name, gene name

# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files, 
                       type    = "kallisto", 
                       tx2gene = tx2gene,
                       countsFromAbundance = "no")

# Save object for use in other scripts
save(txi_counts, file = "results/txi_counts.Rdata")
```

### Step Four: Run DESeq2 for differential expression analysis:

For identifying differentially expressed genes, a DESeq object is generated containing:
    a. sample names
    b. treatments
    
Within the present analysis, three types of questions can be immediately asked:
    1. What genes are differentially expressed across all three treatments?
    2. What genes are differentially expressed within clothianidin-exposed workers in comparison to control?
    3. What genes are differentially expressed within imidacloprid-exposed workers in comparison to control?

```{r}
## Construct DESeq data set from txi object and sample information:
deseq_txi <- DESeqDataSetFromTximport(txi     = txi_counts, 
                                      colData = samples, 
                                      design  = ~ treatment)
```


### Step Five: Run DESeq2 (Wald test) for differential expression analysis:
Which genes are differentially expressed between each treatment and control?
Running a Wald t-test (default test within DESeq2)

```{r}
##======================================================================================
## 5.1) Which genes are differentially expressed within clothianidin-exposed workers in comparison to control?
##======================================================================================
## Perform a pairwise comparison between treatments:
deseq_object_pairwise  <- DESeq(deseq_txi, test = "Wald") #fit model to txi object 

## Extract results for clothianidin vs. control:
deseq_results_pairwise_CLO <- results(deseq_object_pairwise, contrast = c("treatment", "CON", "CLO")) #results of model

## Save object for later use
save(deseq_object_pairwise, file = "deseq_object_pairwise.Rdata")

## lfcShrink(), which performs log2 fold change shrinkage for visualization and ranking of genes:
deseq_results_pairwise_CLO_shrink <- lfcShrink(dds = deseq_object_pairwise, contrast = c("treatment", "CON", "CLO"), res = deseq_results_pairwise_CLO)

# Results table ordered by smallest p value
deseq_results_pairwise_CLO_ordered <- deseq_results_pairwise_CLO_shrink[order(deseq_results_pairwise_CLO_shrink$pvalue),]

# No. genes p < 0.05
deseq_results_pairwise_CLO_significant <- deseq_results_pairwise_CLO_ordered[which(deseq_results_pairwise_CLO_ordered$padj < 0.05), ]

## Log fold changes in expression between clothianidin-exposed workers and control ###
# Genes which are more highly expressed in CLO group, compared to control 
deseq_pairwise_CLO_lfc_high <- deseq_results_pairwise_CLO_significant[which(deseq_results_pairwise_CLO_significant$log2FoldChange < 0),] # n = 21

# Genes which are less expressed in CLO group, compared to control
deseq_pairwise_CLO_lfc_low <- deseq_results_pairwise_CLO_significant[which(deseq_results_pairwise_CLO_significant$log2FoldChange > 0),] # n = 12

# Save results object for use in heatmap
de_genes_pairwise_CLO <- rownames(deseq_results_pairwise_CLO_significant)
save(de_genes_pairwise_CLO, file = "de_genes_pairwise_CLO.Rdata")

# Export gene names object into .txt file, for GO analysis genes of interest 
write(de_genes_pairwise_CLO, "de_genes_pairwise_CLO.txt")
```

##=======================================================================================
## 5.2 Which genes are differentially expressed within imidacloprid-exposed workers in comparison to control?
##=======================================================================================

```{r}
## Perform a pairwise comparison between treatments
deseq_results_pairwise_IMI <- results(deseq_object_pairwise, contrast = c("treatment", "CON", "IMI")) #results of model

#lfcShrink(), which performs log2 fold change shrinkage for visualization and ranking of genes:
deseq_results_pairwise_IMI_shrink <- lfcShrink(dds = deseq_object_pairwise, contrast = c("treatment", "CON", "IMI"), res = deseq_results_pairwise_IMI)

# Results table ordered by smallest p value
deseq_results_pairwise_IMI_ordered <- deseq_results_pairwise_IMI_shrink[order(deseq_results_pairwise_IMI_shrink$pvalue),]

# No. genes p < 0.05
deseq_results_pairwise_IMI_significant <- deseq_results_pairwise_IMI_ordered[which(deseq_results_pairwise_IMI_ordered$padj < 0.05), ] # n = 1
```
