---
title: "dexseq_analysis"
author: "JoeColgan"
date: "14 May 2018"
output: html_document
---

## Introduction
This script is for differential exon usage analysis of exon quantification data. Input files should be a count file (.fb.txt) generated by [HTSeq]("http://htseq.readthedocs.io/en/release_0.10.0/") for each sample and an annotation file (gff) generated by ```dexseq_prepare_annotation.py```, a python script provided with [DEXSeq]("https://bioconductor.org/packages/release/bioc/html/DEXSeq.html").

This script will import count files from HTSeq and use the R package DEXSeq to test for differential exon usage. This script produces a results table of genes that are significantly differentially expressed. This script uses the LRT test in DEXSeq to test for differential exon usage between treatments, which is the only option provided for analysis. Unlike DESeq2, there is no Wald test comparison. 

```{r}
## Load libraries:
libraries <- c("DEXSeq", "ggplot2", "ggpubr")
for (lib in libraries) {
        if (require(package = lib, character.only = TRUE)) {
                print("Successful")
        } else {
                print("Installing")
                source("https://bioconductor.org/biocLite.R")
                biocLite(pkgs = lib)
                library(lib, character.only = TRUE )
        }
}

```
Step One: Read in HTSeq count files
We used HTSeq to extract read counts for each exon bin. 

```{r}
## Read in files:
countFiles = list.files("./input/alternative_splicing_input/clothianidin/", 
                        pattern="fb_counts.txt$", 
                        full.names=TRUE)
basename(countFiles)
countFiles <- countFiles[1:8]## Subset samples of interest

flattenedFile = list.files("./input/alternative_splicing_input/ref_files/", 
                        pattern="gff$", 
                        full.names=TRUE)
basename(flattenedFile)

# Automatically extract file names from kallisto output for colnames
filenames_tmp <- basename(countFiles)
filenames_tmp <- gsub(x = filenames_tmp, pattern = "2016-Bter-", replacement = "")
filenames_tmp <- gsub(x = filenames_tmp, pattern = "-head.+", replacement = "")
filenames_tmp <- gsub(x = filenames_tmp, pattern = ".*Bter-", replacement = "")
## DEXSeq complains about characters, such as '-' so need to replace:
filenames     <- gsub(x = filenames_tmp, pattern = "-", replacement = "_")
```

Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
#samples_tmp <- gsub(x = basename(countFiles), pattern = ".*Bter-", replacement = "")
#samples     <- gsub(x = samples_tmp,          pattern = "-head.*", replacement = "")
#samples     <- data.frame(treatment = samples)
#rownames(samples) <- filenames
#samples$treatment <- gsub(x = samples$treatment ,     pattern = "-.*",     replacement = "")
#samples$treatment <- as.factor(samples$treatment)
#samples$treatment <- relevel(x = samples$treatment, ref = "CON") 
```

Step Three: Create a dexseq object for differential exon usage analysis
We create a dexseq object using:
- The count files (reads aligned per exon) generated by HTSeq.
- The sampleData file outlining the specific treatment each sample is assigned to.
- The variables compared within the GLM used DEXSeq.
- The "flattened" annotation file (gff) generated by the python script () provided with DEXSeq. 

```{r}
## Suppress start up messages as recommended in DEXSeq walkthrough:
suppressPackageStartupMessages( library( "DEXSeq" ) )

## Read in batch information:
samples_information <- read.table(file = "./input/sample_information/room_batch_information_per_sample.txt", header= FALSE)

## Update header information:
colnames(samples_information) <- c("sample_name", "treatment", "room", "batch")
rownames(samples_information) <- samples_information$sample_name

## Convert to factors:
samples_information$batch <- as.factor(samples_information$batch)
samples_information$room <- as.factor(samples_information$room)
samples_information$sample_name <- NULL
samples_information <- samples_information[1:8,]

## Use safer characters:
rownames(samples_information) <- gsub("-", "_", rownames(samples_information))

## Make sure 'CON' is the reference:
samples_information$treatment <- relevel(x = samples_information$treatment, ref = "CON") 

## Create models for testing:
formulaFullModel= ~sample+exon+batch:exon+room:exon+treatment:exon
formulaReducedModel= ~sample+exon+batch:exon+room:exon

## Create DEXSeq object: 
dexseq_object = DEXSeqDataSetFromHTSeq(
                        countFiles,
                        sampleData=samples,
                        design= ~ sample + exon + treatment:exon,
                        flattenedfile=flattenedFile)
```

Step Four: Normalisation of read counts per exon based on library sizes. 

```{r}
## Normalisation of exon counts by library sizes:
dexseq_object = estimateSizeFactors(dexseq_object)
```

Step Five: Dispersion estimation. 
DEXSeq uses the same model as DESeq2 to investigate the dispersion of expression values within replicates per treatment before comparing across values.

```{r}
## Estimate dispersion values:
dexseq_object = estimateDispersions(dexseq_object, 
                                    formula = formulaFullModel)
```

Step Six: Testing for differential exon usage

```{r}
## Test for differential exon usage between treatment groups:
dexseq_object = testForDEU(dexseq_object,
                           reducedModel = formulaReducedModel,
                           fullModel = formulaFullModel)

## Estimation of fold changes in exon usage between treatments:
dexseq_object = estimateExonFoldChanges(dexseq_object, 
                                        fitExpToVar="treatment")

## Subset results:
dexseq_object_results = DEXSeqResults(dexseq_object)
dexseq_object_results_df <- as.data.frame(dexseq_object_results)

## Create a directory for output:
dir.create(path = "./results/alternative_splicing_clo")

## Write results to file:
write.table(dexseq_object_results_df,
            file="./results/alternative_splicing_clo/dexseq_object_results_all_clothianidin_worker_batch_room_corrected.txt",
            row.names = FALSE,
            col.names = TRUE,
            sep="\t",
            quote = FALSE)
            
## Subset columns of interest:
## Gene names plus p-values of significance:
dexseq_object_results_subset <- as.data.frame(cbind(rownames(dexseq_object_results), dexseq_object_results$pvalue))

## Updated column names:
colnames(dexseq_object_results_subset) <- c("locus", "pvalue")

## Output all results:
write.table(dexseq_object_results_subset,
            file="results/alternative_splicing_clo/dexseq_object_results_all_clothianidin.worker__batch_room_exon_raw_pvalues.txt",
            row.names = FALSE,
            col.names = TRUE,
            sep="\t",
            quote = FALSE)

## Investigate how many exons are significantly differently used between treatments
table(dexseq_object_results$padj < 0.05)

## Subset significant genes:
dexseq_object_results_significant <- subset(dexseq_object_results, padj < 0.05)

## Order by signficance:
dexseq_object_results_significant_reordered <- dexseq_object_results_significant[order(dexseq_object_results_significant$padj),]

## Write signifance table to file:
write.table(dexseq_object_results_significant_reordered,
            file="./results/alternative_splicing_clo/dexseq_object_results_sig_clothianidin.txt",
            row.names = TRUE,
            col.names = TRUE,
            sep="\t")
```

Step Seven: Visualisation

```{r}
# M-A plot:
plotMA(dexseq_object_results, 
       cex=0.8, 
       alpha=0.05, 
       ylim=c(-5, 5))

# For examination of DEU for individual genes, use:
plotDEXSeq(dexseq_object_results, 
           geneID=head(dexseq_object_results_significant_reordered$groupID, n=1), 
           fitExpToVar="treatment",
           norCounts = TRUE,
           displayTranscripts = TRUE,
           legend=TRUE, 
           cex.axis=1.2, 
           cex=1.3, 
           lwd=2 )
```
