--- 
title: "Bombus pesticide exposure: gene expression in head"
output: gene_level_3treatments.html
--- 

# Introduction
This script is for differential gene expression analysis of transcript quantification data.
Input files should be abundance files (.h5, .tsv, .json) generated by [Kallisto](https://pachterlab.github.io/kallisto/manual) 
for each sample and .txt file of genome transcripts 
and their corresponding gene ids (LOC numbers)/.

This script will import abundance files from Kallisto and use the R package [DESeq2](http://bioconductor.org/packages/release/bioc/html/DESeq.html) on count data to test for 
differential expression based on a model using negative binomial distribution.
This script produces a results table of genes that are significantly differentially expressed. 
This script uses the Wald test in DESeq2 to test for differential expression between treatments.
This can then be used for further analysis. 


```{r, message = FALSE}
# Load libraries; install from scratch if needed
libraries <- c("tximport", "readr", "DESeq2", "ggplot2", "gplots", "rhdf5",
               "plyr", "ggfortify", "biomaRt", "vsn", "RColorBrewer", "ReportingTools")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
#        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
## Create output directory:
dir.create("results")
```

## Step One: Input results from kallisto

We use raw, non-normalised estimated counts.
```{r} 
## Set paths to folders containing output files from kallisto
paths <- list()
paths$sample_batch_info <- "input/room_batch_information_per_sample-clean.combined.reordered.txt"
paths$kallisto_output   <- "input"

# Set relative paths:
paths$kallisto_files_relative <- grep(x       = list.files(paths$kallisto_output, recursive = TRUE),
                                      pattern = ".h5", 
                                      value   = TRUE)
paths$kallisto_files <- file.path(paths$kallisto_output,
                                  paths$kallisto_files_relative)

# Automatically extract file names from kallisto output for colnames
names(paths$kallisto_files) <- gsub(paths$kallisto_files_relative, pattern = "/.*", replacement = "") 

for (filenumber in 1:length(paths$kallisto_files)) {
  current_name <- names(paths$kallisto_files)[filenumber]
  current_file <- paths$kallisto_files[filenumber]
  if (FALSE == grepl(pattern = current_name, x = current_file)) {
    kill("we have a problem - names and filenames dont match up")
  }
}
```


## Step Two: Put sample names and treatments into a samples dataframe for DESeqDataSetFromTximport

```{r}
## Extract sample names and put into df for tximport
samples     <- data.frame(treatment = names(paths$kallisto_files))

## Read in sample information for room and batch:
samples_information <- read.table(file = paths$sample_batch_info,
                                  header = FALSE, 
                                  col.names = c("sample_name", "treatment", "room", "batch", "caste", "colony"),
                                  row.names = 1)

## Make sure 'CON' is the reference:
samples_information$treatment <- relevel(x = samples_information$treatment, ref = "CON") 

# sanity check
if (FALSE == all(samples$treatment %in% rownames(samples_information))) {
    kill("we have a problem - sample names from files and from batch descriptor file  dont match up")
}
samples_information
```


## Step Three: Import kallisto quantification files using Tximport

We used kallisto to obtain estimated counts for each transcript in each condition. 
We load these raw counts into R for DEseq using tximport. 
Because we do a gene-level analysis first, transcript abundances are summarised per gene during import.

```{r}
# Read in file corresponding to transcripts to gene ids
ensembl <- useMart("metazoa_mart", host = "metazoa.ensembl.org")
ensembl <- useDataset("bterrestris_eg_gene", mart = ensembl)
transcript_to_gene <- getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id"), 
                            mart = ensembl)

# Use tximport on kallisto files 
## Counts are estimated counts from kallisto:
txi_counts <- tximport(paths$kallisto_files, 
                       type    = "kallisto", 
                       tx2gene = transcript_to_gene,
                       countsFromAbundance = "no")

# Save object for use in other scripts
dir.create(path = "results/count_estimates/")
save(txi_counts, file = "results/count_estimates/txi_count_estimates.Rdata")
```

### Step Four: Run DESeq2 for differential expression analysis:

For identifying differentially expressed genes, a DESeq object is generated containing:
    a. sample names
    b. treatments
    
Within the present analysis, two questions can be immediately asked:
    1. What genes are differentially expressed within clothianidin-exposed workers in comparison to control?
    2. What genes are differentially expressed within imidacloprid-exposed workers in comparison to control?

# Run DESeq2 (Wald test) for differential expression analysis:
  Which genes are differentially expressed between each treatment and control?
  Running a Wald t-test (default test within DESeq2)

## Set reference caste equal to workers first:

```{r, message=FALSE}
## Set workers as reference for caste:
samples_information$caste <- relevel(x = samples_information$caste, ref = "workers") 

## Rerun model design:
deseq_txi <- DESeqDataSetFromTximport(txi     = txi_counts, 
                                      colData = samples_information, 
                                      design  = ~batch + caste + treatment + caste:treatment)

## Perform a pairwise comparison between treatments:
deseq_object  <- DESeq(deseq_txi, 
                       test = "Wald", 
                       betaPrior = FALSE)

group_names <- grep("treatment_", resultsNames(deseq_object), value=TRUE)
treatments <- gsub("treatment_", "", group_names)
treatments <- gsub("_vs_CON", "", treatments)
## Set control treatment:
control <- "CON"
non_control_treatments <- treatments[!treatments %in% control]

## So the current model does not use betaPrior meaning that spurious results may be introduced:
## First of all we want to check the condition effect for the reference caste (i.e. queens)
for (pesticide in non_control_treatments){
        ## print pesticide to console:
        print(pesticide)
        ## Contrast pesticide and control:
        results_tmp <- results(deseq_object, contrast=c("treatment", pesticide, control))
        ## Convert to dataframe and sort by significance (adjusted p-value):
        results_tmp_df <- as.data.frame(results_tmp)
        ## Reorder by significance and count number of significantly differentially expressed gene:
        results_tmp_df_sorted <- results_tmp_df[order(results_tmp_df$padj), ]
        results_tmp_df_sorted_genes <- nrow(subset(results_tmp_df_sorted, padj < 0.05))
        print(c(paste("Main effect:", results_tmp_df_sorted_genes)))
        ## Extract results for 
        interaction <- c(paste("castequeens.treatment", pesticide, sep=""))
        results_tmp_2 <- results(deseq_object, 
                                 list( c(paste("treatment_", pesticide, "_vs_", control, sep = "")),
                                         c(interaction) ))
        ## Convert to dataframe, sort and count number of significant DEGs:
        results_tmp_2_df <- as.data.frame(results_tmp_2)
        results_tmp_2_df_sorted <- results_tmp_2_df[order(results_tmp_2_df$padj), ]
        results_tmp_2_df_sorted_genes <- nrow(subset(results_tmp_2_df_sorted, padj < 0.05))
        print(c(paste("Main effect plus interaction effect:", results_tmp_2_df_sorted_genes)))
        ## Step Three: investigate whether treatments effects castes differently:
        diff_across_castes <- results(deseq_object, name=interaction)
        diff_across_castes_df <- as.data.frame(diff_across_castes)
        diff_across_castes_df_sorted <- diff_across_castes_df[order(diff_across_castes_df$padj), ]
        diff_across_castes_df_sorted_genes <- nrow(subset(diff_across_castes_df_sorted, padj < 0.05))
        print(c(paste("Total number of genes:", diff_across_castes_df_sorted_genes)))
}

```
Run analysis for both pesticide treatments using queen as reference caste. 

```{r, message=FALSE}
## Set queens as reference for caste:
samples_information$caste <- relevel(x = samples_information$caste, ref = "queens") 

## Rerun model design:
deseq_txi <- DESeqDataSetFromTximport(txi     = txi_counts, 
                                      colData = samples_information, 
                                      design  = ~batch + caste + treatment + caste:treatment)

## Perform a pairwise comparison between treatments:
deseq_object  <- DESeq(deseq_txi, 
                       test = "Wald", 
                       betaPrior = FALSE)

group_names <- grep("treatment_", resultsNames(deseq_object), value=TRUE)
treatments <- gsub("treatment_", "", group_names)
treatments <- gsub("_vs_CON", "", treatments)
## Set control treatment:
control <- "CON"
non_control_treatments <- treatments[!treatments %in% control]

## So the current model does not use betaPrior meaning that spurious results may be introduced:
## First of all we want to check the condition effect for the reference caste (i.e. queens)
for (pesticide in non_control_treatments){
        ## print pesticide to console:
        print(pesticide)
        ## Step One: Condition effect for caste (the main effect)
        ## Contrast pesticide and control:
        results_tmp <- results(deseq_object, contrast=c("treatment", pesticide, control))
        ## Convert to dataframe and sort by significance (adjusted p-value):
        results_tmp_df <- as.data.frame(results_tmp)
        ## Reorder by significance and count number of significantly differentially expressed gene:
        results_tmp_df_sorted <- results_tmp_df[order(results_tmp_df$padj), ]
        results_tmp_df_sorted_genes <- nrow(subset(results_tmp_df_sorted, padj < 0.05))
        print(c(paste("Main effect:", results_tmp_df_sorted_genes)))
        ## Step Two: Extract results for main effect of condition plus the interaction term:
        interaction <- c(paste("casteworkers.treatment", pesticide, sep=""))
        results_tmp_2 <- results(deseq_object, 
                                 list( c(paste("treatment_", pesticide, "_vs_", control, sep = "")),
                                         c(interaction) ))
        ## Convert to dataframe, sort and count number of significant DEGs:
        results_tmp_2_df <- as.data.frame(results_tmp_2)
        results_tmp_2_df_sorted <- results_tmp_2_df[order(results_tmp_2_df$padj), ]
        results_tmp_2_df_sorted_genes <- nrow(subset(results_tmp_2_df_sorted, padj < 0.05))
        print(c(paste("Main effect plus interaction effect:", results_tmp_2_df_sorted_genes)))
        ## Step Three: investigate whether treatments effects castes differently:
        diff_across_castes <- results(deseq_object, name=interaction)
        diff_across_castes_df <- as.data.frame(diff_across_castes)
        diff_across_castes_df_sorted <- diff_across_castes_df[order(diff_across_castes_df$padj), ]
        diff_across_castes_df_sorted_genes <- nrow(subset(diff_across_castes_df_sorted, padj < 0.05))
        print(c(paste("Total number of genes:", diff_across_castes_df_sorted_genes)))
}
```

Generate a PCA using all samples:

```{r}
## Extracting transformed values from DESeq object:
vsd <- vst(deseq_object, blind = FALSE)
rld <- rlog(deseq_object, blind = FALSE)

# this gives log2(n + 1)
ntd        <- normTransform(deseq_object)
## Basic plot:
plotPCA(vsd, intgroup = c("treatment"))
## Extract first and second principal components:
pca_data    <- plotPCA(vsd, intgroup = c("treatment"), returnData = TRUE)
## Extract percentages for first two principal components:
percentVar <- round(100 * attr(pca_data, "percentVar"))

## Plot PCA:
ggplot(pca_data, aes(PC1, PC2, color = treatment, fill = treatment)) +
       geom_point(size = 5, shape = 16) +
       xlab(paste0("PC1: ",percentVar[1],"% variance")) +
       ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
       coord_fixed() +
       scale_color_manual(values = c("blue", "grey", "black")) +
       theme_bw() +
       geom_text(label = pca_data$name) +
       theme(axis.title = element_text(family = "Arial", color = "#666666", face = "bold", size = 16)) 
    
```
