--- 
title: "Graphical representation of DE genes"
output: 2018-02-28-graphics.nb.html
--- 
---

## Introduction:
#=================
##This script is for exploratory analysis of transcript quantification data and results of differential gene expression analysis.
##Input files should be R objects saved and generated from previous scripts of DE analysis, including the object resulting from using tximport() on kallisto files, the significant transcript abundances selected from this object and a results table of significant genes, generated using DESeq2. 
##This script will organise input data for generating heatmaps of significantly differentially expressed genes, based on statistical significance and log fold changes in expression.
##This script produces heatmaps that can be exported. 

```{r}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2", "gplots", "heatmap3", "ggfortify", "ggDESeq")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}

# Load R objects from previous script output
load("txi_counts.Rdata")
load("deseq_object_LRT.Rdata")
load("deseq_object_pairwise.Rdata")

```


## Step One: Heatmaps for visualisation of significantly differentially expressed transcripts
```{r}
## Generate results table for DE genes between treatments:
# Set colour pattern
scalepurpleyellow <- colorRampPalette(colors = c("black", "white", "red"))(400)

## The function takes a z-normalised dataframe as input
plot_heatmap3 <- function(input){
                  heatmap.2(input, 
                  col = scalepurpleyellow, 
                  trace = "none",
                  margins = c(10,14), 
                  sepwidth = c(0.01, 0.01), 
                  sepcolor = "black", 
                  colsep = 1:ncol(input),
                  rowsep = 1:nrow(input),
                  cexRow = 1,
                  Colv = NULL,
                  labRow = rownames(input))
}
  
###### a) Heatmap of all three treatments ######
## Select relevant gene counts from DESeq results (de_genes)
# normalise counts 
normalised_counts_df <- t(scale(t(txi_counts$counts), center = TRUE, scale = TRUE))

# Select significant gene counts?
significant_de_counts <- normalised_counts_df[de_genes,]
# Add sample numbers to columns 
colnames(significant_de_counts) <- gsub(x = colnames(significant_de_counts), pattern = "-C.*", replacement = "")
colnames(significant_de_counts) <- paste(colnames(significant_de_counts),+rep(1:4,2))

# Plot
plot_heatmap3(significant_de_counts)
  
  

###### b) Heatmap of pairwise comparisons between treatments ######

## Create function to generate results tables for pairwise comparisons:
generate_pairwise_heatmap <- function(deseq_input, treatment_1, treatment_2, counts_input){
                          ## Shrink log fold changes:
                          deseq_results_shrink <- lfcShrink(deseq_input, contrast = c("treatment", treatment_1, treatment_2))
                          ## Reorder results:
                          deseq_results_shrink_ordered <- deseq_results_shrink[order(deseq_results_shrink$padj),]
                          ## Subset significant values:
                          de_genes <- deseq_results_shrink_ordered[which(deseq_results_shrink_ordered$padj < 0.05), ]
                          ## Print to console:
                          print(nrow(de_genes))
                          # normalise counts 
                          normalised_counts_df <- t(scale(t(counts_input), center = TRUE, scale = TRUE))
                          # Select significant gene counts
                          significant_de_counts <- normalised_counts_df[row.names(de_genes),]
                          ## Return:
                          return(plot_heatmap3(significant_de_counts))
}

## Call function:
generate_pairwise_heatmap(deseq_object_LRT, "CON", "CLO", txi_counts$counts) 
                          

```









