---
title: "Bombus pesticide exposure: gene expression in head"
output: run_kallisto_hisat2_differential_expression_comparison.html
author: Joe Colgan, Yannick Wurm http://wurmlab.com
---

## Introduction
This script takes tab-delimited text files as input, which contain the output results of differential expression analyses using the R package, DESeq2. Differential expression analysis was performed by DESeq2 for queens and workers exposed to the neonicotinoids, clothianidin and imidacloprid, using two input data sets for each caste. One data set consisted of estimated counts generated by the pseudoaligner, Kallisto, while the second data set consisted of actual counts generated from genome-level alignments performed by HISAT2 and gene-level read counts extracted using HTSeq. 

The output of this script is the generation of Venn diagram plots.

The first step is to load libraries required for the present analysis. 

```{r, message=FALSE}
# Load libraries; install from scratch if needed
libraries <- c("ggplot2", "lintr", "ggpubr", "hash", "eulerr")
for (lib in libraries) {
    if (require(package = lib, character.only = TRUE)) {
        print("Successful")
    } else {
        print("Installing")
        source("https://bioconductor.org/biocLite.R")
        biocLite(pkgs = lib)
        library(lib, character.only = TRUE )
    }
}
```

## Step One: Load input files:
Read in input files and store as dataframes. 

```{r, message=FALSE}
## Read in input files for queens:
queen_kallisto <- read.table(file = "./input/queen/kallisto_queen_de_genes_sig_for_euler.tmp",
                             header = FALSE,
                             col.names = "ids")$ids
queen_hisat2 <- read.table(file = "./input/queen/hisat2_queen_de_genes_sig_for_euler.tmp",
                           header = FALSE,
                           col.names = "ids")$ids

## Read in input files for workers:
worker_kallisto <- read.table(file = "./input/worker/kallisto_worker_de_genes_sig_for_euler.tmp",
                             header = FALSE,
                             col.names = "ids")$ids 
worker_hisat2 <- read.table(file = "./input/worker/hisat2_worker_de_genes_sig_for_euler.tmp",
                           header = FALSE,
                             col.names = "ids")$ids
```

## Step Two: Create combinations for plotting:

```{r, message = FALSE}
## Combine lists:
venneu_queen <- euler(combinations = list("Kallisto \nQueen" = queen_kallisto,
                                    "HISAT2-HTSeq \nQueen" = queen_hisat2))

venneu_worker <- euler(combinations = list("Kallisto \nWorker" = worker_kallisto,
                                    "HISAT2-HTSeq \nWorker" = worker_hisat2))

```

## Step Three: Generate plot:

```{r, message = FALSE}
## Generate plot:
queen_plots <- plot(venneu_queen,
  quantities = TRUE,
  edges = TRUE,
  labels = list(fontsize = 12))

worker_plots <- plot(venneu_worker,
  quantities = TRUE,
  edges = TRUE,
  labels = list(fontsize = 12))

## Generate a combined plot:
ggarrange(queen_plots, worker_plots,
          labels = c("A) Queen DE genes",
                     "B) Worker DE genes"),
          ncol = 1, nrow = 2)

## Save picture:
ggsave("caste_differential_expression_euler_plot.png",
       height = 10, width = 8)
```